<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="Comprehensive toolkit for generating various numerical
    features of protein sequences described in Xiao et al. (2015)
    &lt;DOI:10.1093/bioinformatics/btv042&gt;. For full functionality,
    the software ncbi-blast+ is needed, see
    &lt;https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download&gt;
    for more information.">
<title>Generating Various Numerical Representation Schemes for Protein Sequences • protr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="apple-touch-icon-60x60.png">
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Generating Various Numerical Representation Schemes for Protein Sequences">
<meta property="og:description" content="Comprehensive toolkit for generating various numerical
    features of protein sequences described in Xiao et al. (2015)
    &lt;DOI:10.1093/bioinformatics/btv042&gt;. For full functionality,
    the software ncbi-blast+ is needed, see
    &lt;https://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&amp;DOC_TYPE=Download&gt;
    for more information.">
<meta property="og:image" content="https://nanx.me/protr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">protr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7-0</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/protr.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nanxstats/protr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header">
<img src="logo.png" class="logo" alt=""><h1 id="protr-">protr <a class="anchor" aria-label="anchor" href="#protr-"></a>
</h1>
</div>
<!-- badges: start -->

<p>Comprehensive toolkit for generating various numerical features of protein sequences described in Xiao et al. (2015) &lt;<a href="https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/btv042" class="external-link">DOI:10.1093/bioinformatics/btv042</a>&gt; (<a href="https://nanx.me/papers/protr.pdf" class="external-link">PDF</a>).</p>
<div class="section level2">
<h2 id="paper-citation">Paper citation<a class="anchor" aria-label="anchor" href="#paper-citation"></a>
</h2>
<p>Formatted citation:</p>
<p>Nan Xiao, Dong-Sheng Cao, Min-Feng Zhu, Qing-Song Xu (2015). protr/ProtrWeb: R package and web server for generating various numerical representation schemes of protein sequences. <em>Bioinformatics</em>, <strong>31</strong>(11), 1857–1859.</p>
<p>BibTeX entry:</p>
<pre><code>@article{Xiao2015,
  author  = {Xiao, Nan and Cao, Dong-Sheng and Zhu, Min-Feng and Xu, Qing-Song.},
  title   = {protr/{ProtrWeb}: {R} package and web server for generating various numerical representation schemes of protein sequences},
  journal = {Bioinformatics},
  year    = {2015},
  volume  = {31},
  number  = {11},
  pages   = {1857--1859},
  doi     = {10.1093/bioinformatics/btv042}
}</code></pre>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install <code>protr</code> from CRAN:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"protr"</span><span class="op">)</span></span></code></pre></div>
<p>Or try the latest version on GitHub:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">remotes</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"nanxstats/protr"</span><span class="op">)</span></span></code></pre></div>
<p><a href="https://nanx.me/protr/articles/protr.html">Browse the package vignette</a> for a quick-start.</p>
</div>
<div class="section level2">
<h2 id="shiny-app">Shiny app<a class="anchor" aria-label="anchor" href="#shiny-app"></a>
</h2>
<p>ProtrWeb, the Shiny web application built on protr, can be accessed from <a href="http://protr.org" class="external-link">http://protr.org</a>.</p>
<p>ProtrWeb is a user-friendly web application for computing the protein sequence descriptors (features) presented in the protr package.</p>
</div>
<div class="section level2">
<h2 id="list-of-supported-descriptors">List of supported descriptors<a class="anchor" aria-label="anchor" href="#list-of-supported-descriptors"></a>
</h2>
<div class="section level3">
<h3 id="commonly-used-descriptors">Commonly used descriptors<a class="anchor" aria-label="anchor" href="#commonly-used-descriptors"></a>
</h3>
<ul>
<li>
<p>Amino acid composition descriptors</p>
<ul>
<li>Amino acid composition</li>
<li>Dipeptide composition</li>
<li>Tripeptide composition</li>
</ul>
</li>
<li>
<p>Autocorrelation descriptors</p>
<ul>
<li>Normalized Moreau-Broto autocorrelation</li>
<li>Moran autocorrelation</li>
<li>Geary autocorrelation</li>
</ul>
</li>
<li>
<p>CTD descriptors</p>
<ul>
<li>Composition</li>
<li>Transition</li>
<li>Distribution</li>
</ul>
</li>
<li><p>Conjoint Triad descriptors</p></li>
<li>
<p>Quasi-sequence-order descriptors</p>
<ul>
<li>Sequence-order-coupling number</li>
<li>Quasi-sequence-order descriptors</li>
</ul>
</li>
<li>
<p>Pseudo amino acid composition (PseAAC)</p>
<ul>
<li>Pseudo amino acid composition</li>
<li>Amphiphilic pseudo amino acid composition</li>
</ul>
</li>
<li>
<p>Profile-based descriptors</p>
<ul>
<li>Profile-based descriptors derived by PSSM (Position-Specific Scoring Matrix)</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="proteochemometric-pcm-modeling-descriptors">Proteochemometric (PCM) modeling descriptors<a class="anchor" aria-label="anchor" href="#proteochemometric-pcm-modeling-descriptors"></a>
</h3>
<ul>
<li>Scales-based descriptors derived by principal components analysis
<ul>
<li>Scales-based descriptors derived by amino acid properties (AAindex)</li>
<li>Scales-based descriptors derived by 20+ classes of 2D and 3D molecular descriptors (Topological, WHIM, VHSE, etc.)</li>
<li>Scales-based descriptors derived by factor analysis</li>
<li>Scales-based descriptors derived by multidimensional scaling</li>
<li>BLOSUM and PAM matrix-derived descriptors</li>
</ul>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="similarity-computation">Similarity computation<a class="anchor" aria-label="anchor" href="#similarity-computation"></a>
</h3>
<p>Local and global pairwise sequence alignment for protein sequences:</p>
<ul>
<li>Between two protein sequences</li>
<li>Parallelized pairwise similarity calculation with a list of protein sequences</li>
<li>Parallelized pairwise similarity calculation between two sets of protein sequences</li>
</ul>
<p>GO semantic similarity measures:</p>
<ul>
<li>Between two groups of GO terms / two Entrez Gene IDs</li>
<li>Parallelized pairwise similarity calculation with a list of GO terms / Entrez Gene IDs</li>
</ul>
</div>
<div class="section level3">
<h3 id="miscellaneous-tools-and-datasets">Miscellaneous tools and datasets<a class="anchor" aria-label="anchor" href="#miscellaneous-tools-and-datasets"></a>
</h3>
<ul>
<li>Retrieve protein sequences from UniProt</li>
<li>Read protein sequences in FASTA format</li>
<li>Read protein sequences in PDB format</li>
<li>Sanity check of the amino acid types appeared in the protein sequences</li>
<li>Protein sequence segmentation</li>
<li>Auto cross covariance (ACC) for generating scales-based descriptors of the same length</li>
<li>20+ pre-computed 2D and 3D descriptor sets for the 20 amino acids to use with the scales-based descriptors</li>
<li>BLOSUM and PAM matrices for the 20 amino acids</li>
<li>Meta information of the 20 amino acids</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="contribute">Contribute<a class="anchor" aria-label="anchor" href="#contribute"></a>
</h2>
<p>To contribute to this project, please take a look at the <a href="https://nanx.me/protr/CONTRIBUTING.html">Contributing Guidelines</a> first. Please note that this project is released with a <a href="https://nanx.me/protr/CONDUCT.html">Contributor Code of Conduct</a>. By participating in this project you agree to abide by its terms.</p>
</div>
</div>
  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=protr" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/nanxstats/protr/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/nanxstats/protr/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/BSD-3-Clause" class="external-link">BSD_3_clause</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>

<div class="community">
<h2 data-toc-skip>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
</ul>
</div>

<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing protr</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Nan Xiao <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-0250-5673" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Qing-Song Xu <br><small class="roles"> Author </small>  </li>
<li>Dong-Sheng Cao <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/nanxstats/protr/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/nanxstats/protr/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://cran.r-project.org/package=protr" class="external-link"><img src="https://www.r-pkg.org/badges/version/protr" alt="CRAN Version"></a></li>
<li><a href="https://cranlogs.r-pkg.org/badges/protr" class="external-link"><img src="https://cranlogs.r-pkg.org/badges/protr" alt="Downloads from the RStudio CRAN mirror"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Nan Xiao, Qing-Song Xu, Dong-Sheng Cao.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
