<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Compute PSSM (Position-Specific Scoring Matrix) for given protein sequence — extractPSSM • protr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Compute PSSM (Position-Specific Scoring Matrix) for given protein sequence — extractPSSM"><meta name="description" content="This function calculates the PSSM (Position-Specific Scoring Matrix) derived
by PSI-Blast for given protein sequence or peptides."><meta property="og:description" content="This function calculates the PSSM (Position-Specific Scoring Matrix) derived
by PSI-Blast for given protein sequence or peptides."><meta property="og:image" content="https://nanx.me/protr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">protr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7-2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/protr.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/nanxstats/protr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Compute PSSM (Position-Specific Scoring Matrix) for given protein sequence</h1>
      <small class="dont-index">Source: <a href="https://github.com/nanxstats/protr/blob/v1.7-2/R/desc-15-PSSM.R" class="external-link"><code>R/desc-15-PSSM.R</code></a></small>
      <div class="d-none name"><code>extractPSSM.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates the PSSM (Position-Specific Scoring Matrix) derived
by PSI-Blast for given protein sequence or peptides.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">extractPSSM</span><span class="op">(</span></span>
<span>  <span class="va">seq</span>,</span>
<span>  start.pos <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  end.pos <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nchar.html" class="external-link">nchar</a></span><span class="op">(</span><span class="va">seq</span><span class="op">)</span>,</span>
<span>  psiblast.path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  makeblastdb.path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  database.path <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  iter <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  silent <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  evalue <span class="op">=</span> <span class="fl">10L</span>,</span>
<span>  word.size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gapopen <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gapextend <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  matrix <span class="op">=</span> <span class="st">"BLOSUM62"</span>,</span>
<span>  threshold <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  seg <span class="op">=</span> <span class="st">"no"</span>,</span>
<span>  soft.masking <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  culling.limit <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  best.hit.overhang <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  best.hit.score.edge <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  xdrop.ungap <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  xdrop.gap <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  xdrop.gap.final <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  window.size <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gap.trigger <span class="op">=</span> <span class="fl">22L</span>,</span>
<span>  num.threads <span class="op">=</span> <span class="fl">1L</span>,</span>
<span>  pseudocount <span class="op">=</span> <span class="fl">0L</span>,</span>
<span>  inclusion.ethresh <span class="op">=</span> <span class="fl">0.002</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-seq">seq<a class="anchor" aria-label="anchor" href="#arg-seq"></a></dt>
<dd><p>Character vector, as the input protein sequence.</p></dd>


<dt id="arg-start-pos">start.pos<a class="anchor" aria-label="anchor" href="#arg-start-pos"></a></dt>
<dd><p>Optional integer denoting the start position of the
fragment window. Default is <code>1</code>,
i.e. the first amino acid of the given sequence.</p></dd>


<dt id="arg-end-pos">end.pos<a class="anchor" aria-label="anchor" href="#arg-end-pos"></a></dt>
<dd><p>Optional integer denoting the end position of the
fragment window. Default is <code>nchar(seq)</code>,
i.e. the last amino acid of the given sequence.</p></dd>


<dt id="arg-psiblast-path">psiblast.path<a class="anchor" aria-label="anchor" href="#arg-psiblast-path"></a></dt>
<dd><p>Character string indicating the path of the
<code>psiblast</code> program.
If NCBI Blast+ was previously installed in the operation system,
the path will be automatically detected.</p></dd>


<dt id="arg-makeblastdb-path">makeblastdb.path<a class="anchor" aria-label="anchor" href="#arg-makeblastdb-path"></a></dt>
<dd><p>Character string indicating the path of the
<code>makeblastdb</code> program.
If NCBI Blast+ was previously installed in the system,
the path will be automatically detected.</p></dd>


<dt id="arg-database-path">database.path<a class="anchor" aria-label="anchor" href="#arg-database-path"></a></dt>
<dd><p>Character string indicating the path of
a reference database (a FASTA file).</p></dd>


<dt id="arg-iter">iter<a class="anchor" aria-label="anchor" href="#arg-iter"></a></dt>
<dd><p>Number of iterations to perform for PSI-Blast.</p></dd>


<dt id="arg-silent">silent<a class="anchor" aria-label="anchor" href="#arg-silent"></a></dt>
<dd><p>Logical. Whether the PSI-Blast running output
should be shown or not (May not work on some Windows versions and
PSI-Blast versions), default is <code>TRUE</code>.</p></dd>


<dt id="arg-evalue">evalue<a class="anchor" aria-label="anchor" href="#arg-evalue"></a></dt>
<dd><p>Expectation value (E) threshold for saving hits.
Default is <code>10</code>.</p></dd>


<dt id="arg-word-size">word.size<a class="anchor" aria-label="anchor" href="#arg-word-size"></a></dt>
<dd><p>Word size for wordfinder algorithm. An integer &gt;= 2.</p></dd>


<dt id="arg-gapopen">gapopen<a class="anchor" aria-label="anchor" href="#arg-gapopen"></a></dt>
<dd><p>Integer. Cost to open a gap.</p></dd>


<dt id="arg-gapextend">gapextend<a class="anchor" aria-label="anchor" href="#arg-gapextend"></a></dt>
<dd><p>Integer. Cost to extend a gap.</p></dd>


<dt id="arg-matrix">matrix<a class="anchor" aria-label="anchor" href="#arg-matrix"></a></dt>
<dd><p>Character string. The scoring matrix name
(default is <code>"BLOSUM62"</code>).</p></dd>


<dt id="arg-threshold">threshold<a class="anchor" aria-label="anchor" href="#arg-threshold"></a></dt>
<dd><p>Minimum word score such that the word is added to
the BLAST lookup table. A real value &gt;= 0.</p></dd>


<dt id="arg-seg">seg<a class="anchor" aria-label="anchor" href="#arg-seg"></a></dt>
<dd><p>Character string. Filter query sequence with SEG (<code>"yes"</code>,
<code>"window locut hicut"</code>, or <code>"no"</code> to disable).
Default is <code>"no"</code>.</p></dd>


<dt id="arg-soft-masking">soft.masking<a class="anchor" aria-label="anchor" href="#arg-soft-masking"></a></dt>
<dd><p>Logical. Apply filtering locations as soft masks?
Default is <code>FALSE</code>.</p></dd>


<dt id="arg-culling-limit">culling.limit<a class="anchor" aria-label="anchor" href="#arg-culling-limit"></a></dt>
<dd><p>An integer &gt;= 0. If the query range of a hit is
enveloped by that of at least this many higher-scoring hits,
delete the hit. Incompatible with <code>best.hit.overhang</code> and
<code>best_hit_score_edge</code>.</p></dd>


<dt id="arg-best-hit-overhang">best.hit.overhang<a class="anchor" aria-label="anchor" href="#arg-best-hit-overhang"></a></dt>
<dd><p>Best Hit algorithm overhang value
(A real value &gt;= 0 and =&lt; 0.5, recommended value: 0.1).
Incompatible with <code>culling_limit</code>.</p></dd>


<dt id="arg-best-hit-score-edge">best.hit.score.edge<a class="anchor" aria-label="anchor" href="#arg-best-hit-score-edge"></a></dt>
<dd><p>Best Hit algorithm score edge value
(A real value &gt;=0 and =&lt; 0.5, recommended value: 0.1).
Incompatible with <code>culling_limit</code>.</p></dd>


<dt id="arg-xdrop-ungap">xdrop.ungap<a class="anchor" aria-label="anchor" href="#arg-xdrop-ungap"></a></dt>
<dd><p>X-dropoff value (in bits) for ungapped extensions.</p></dd>


<dt id="arg-xdrop-gap">xdrop.gap<a class="anchor" aria-label="anchor" href="#arg-xdrop-gap"></a></dt>
<dd><p>X-dropoff value (in bits) for preliminary gapped extensions.</p></dd>


<dt id="arg-xdrop-gap-final">xdrop.gap.final<a class="anchor" aria-label="anchor" href="#arg-xdrop-gap-final"></a></dt>
<dd><p>X-dropoff value (in bits) for final gapped alignment.</p></dd>


<dt id="arg-window-size">window.size<a class="anchor" aria-label="anchor" href="#arg-window-size"></a></dt>
<dd><p>An integer &gt;= 0. Multiple hits window size,
To specify 1-hit algorithm, use <code>0</code>.</p></dd>


<dt id="arg-gap-trigger">gap.trigger<a class="anchor" aria-label="anchor" href="#arg-gap-trigger"></a></dt>
<dd><p>Number of bits to trigger gapping. Default is <code>22</code>.</p></dd>


<dt id="arg-num-threads">num.threads<a class="anchor" aria-label="anchor" href="#arg-num-threads"></a></dt>
<dd><p>Integer. Number of threads (CPUs) to use in the
BLAST search. Default is <code>1</code>.</p></dd>


<dt id="arg-pseudocount">pseudocount<a class="anchor" aria-label="anchor" href="#arg-pseudocount"></a></dt>
<dd><p>Integer. Pseudo-count value used when constructing PSSM.
Default is <code>0</code>.</p></dd>


<dt id="arg-inclusion-ethresh">inclusion.ethresh<a class="anchor" aria-label="anchor" href="#arg-inclusion-ethresh"></a></dt>
<dd><p>E-value inclusion threshold for pairwise alignments.
Default is <code>0.002</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>The original PSSM, a numeric matrix which has
<code>end.pos - start.pos + 1</code> columns and <code>20</code> named rows.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>For given protein sequences or peptides, PSSM represents the
log-likelihood of the substitution of the 20 types of amino acids at that
position in the sequence. Note that the output value is not normalized.</p>
    </div>
    <div class="section level2">
    <h2 id="note">Note<a class="anchor" aria-label="anchor" href="#note"></a></h2>
    <p>The function requires the <code>makeblastdb</code> and <code>psiblast</code> programs
to be properly installed in the operation system or
their paths provided.</p>
<p>The two command-line programs are included in the NCBI-BLAST+
software package. To install NCBI Blast+ for your operating system, see
<a href="https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html" class="external-link">https://blast.ncbi.nlm.nih.gov/doc/blast-help/downloadblastdata.html</a>
for detailed instructions.</p>
<p>Ubuntu/Debian users can directly use the command
<code>sudo apt-get install ncbi-blast+</code> to install NCBI Blast+.
For OS X users, download <code>ncbi-blast- ... .dmg</code> then install.
For Windows users, download <code>ncbi-blast- ... .exe</code> then install.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Altschul, Stephen F., et al.
"Gapped BLAST and PSI-BLAST: a new generation of protein database search programs."
<em>Nucleic acids research</em> 25.17 (1997): 3389–3402.</p>
<p>Ye, Xugang, Guoli Wang, and Stephen F. Altschul.
"An assessment of substitution scores for protein profile-profile comparison."
<em>Bioinformatics</em> 27.24 (2011): 3356–3363.</p>
<p>Rangwala, Huzefa, and George Karypis.
"Profile-based direct kernels for remote homology detection and fold recognition."
<em>Bioinformatics</em> 21.23 (2005): 4239–4247.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><a href="extractPSSMFeature.html">extractPSSMFeature</a> <a href="extractPSSMAcc.html">extractPSSMAcc</a></p></div>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Nan Xiao &lt;<a href="https://nanx.me" class="external-link">https://nanx.me</a>&gt;</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.which.html" class="external-link">Sys.which</a></span><span class="op">(</span><span class="st">"makeblastdb"</span><span class="op">)</span> <span class="op">==</span> <span class="st">""</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.which.html" class="external-link">Sys.which</a></span><span class="op">(</span><span class="st">"psiblast"</span><span class="op">)</span> <span class="op">==</span> <span class="st">""</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="st">"Cannot find makeblastdb or psiblast. Please install NCBI Blast+ first"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span></span>
<span class="r-in"><span>  <span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"protseq/P00750.fasta"</span>,</span></span>
<span class="r-in"><span>    package <span class="op">=</span> <span class="st">"protr"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span>  <span class="va">dbpath</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="st">"tempdb"</span>, fileext <span class="op">=</span> <span class="st">".fasta"</span><span class="op">)</span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/files.html" class="external-link">file.copy</a></span><span class="op">(</span>from <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"protseq/Plasminogen.fasta"</span>,</span></span>
<span class="r-in"><span>    package <span class="op">=</span> <span class="st">"protr"</span></span></span>
<span class="r-in"><span>  <span class="op">)</span>, to <span class="op">=</span> <span class="va">dbpath</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="va">pssmmat</span> <span class="op">&lt;-</span> <span class="fu">extractPSSM</span><span class="op">(</span>seq <span class="op">=</span> <span class="va">x</span>, database.path <span class="op">=</span> <span class="va">dbpath</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span>  <span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">pssmmat</span><span class="op">)</span> <span class="co"># 20 x 562 (P00750: length 562, 20 Amino Acids)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cannot find makeblastdb or psiblast. Please install NCBI Blast+ first</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nan Xiao, Qing-Song Xu, Dong-Sheng Cao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

