<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Scales-Based Descriptors derived by Factor Analysis — extractFAScales • protr</title><!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png"><link rel="icon" type="”image/svg+xml”" href="../favicon.svg"><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" sizes="any" href="../favicon.ico"><link rel="manifest" href="../site.webmanifest"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Scales-Based Descriptors derived by Factor Analysis — extractFAScales"><meta name="description" content="This function calculates scales-based descriptors
derived by Factor Analysis (FA).
Users can provide customized amino acid property matrices."><meta property="og:description" content="This function calculates scales-based descriptors
derived by Factor Analysis (FA).
Users can provide customized amino acid property matrices."><meta property="og:image" content="https://nanx.me/protr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">protr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7-5</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/protr.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/nanxstats/protr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Scales-Based Descriptors derived by Factor Analysis</h1>
      <small class="dont-index">Source: <a href="https://github.com/nanxstats/protr/blob/v1.7-5/R/pcm-04-extractFAScales.R" class="external-link"><code>R/pcm-04-extractFAScales.R</code></a></small>
      <div class="d-none name"><code>extractFAScales.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates scales-based descriptors
derived by Factor Analysis (FA).
Users can provide customized amino acid property matrices.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">extractFAScales</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">propmat</span>,</span>
<span>  <span class="va">factors</span>,</span>
<span>  scores <span class="op">=</span> <span class="st">"regression"</span>,</span>
<span>  <span class="va">lag</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  silent <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A character vector, as the input protein sequence.</p></dd>


<dt id="arg-propmat">propmat<a class="anchor" aria-label="anchor" href="#arg-propmat"></a></dt>
<dd><p>A matrix containing the properties for the amino acids.
Each row represent one amino acid type, each column represents one property.
Note that the one-letter row names must be provided for we need them to seek
the properties for each AA type.</p></dd>


<dt id="arg-factors">factors<a class="anchor" aria-label="anchor" href="#arg-factors"></a></dt>
<dd><p>Integer. The number of factors to be fitted.
Must be no greater than the number of AA properties provided.</p></dd>


<dt id="arg-scores">scores<a class="anchor" aria-label="anchor" href="#arg-scores"></a></dt>
<dd><p>Type of scores to produce. The default is <code>"regression"</code>,
which gives Thompson's scores, <code>"Bartlett"</code> given Bartlett's weighted
least-squares scores.</p></dd>


<dt id="arg-lag">lag<a class="anchor" aria-label="anchor" href="#arg-lag"></a></dt>
<dd><p>The lag parameter. Must be less than the amino acids number
in the protein sequence.</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>Logical. Should we auto-scale the property matrix
(<code>propmat</code>) before doing Factor Analysis? Default is <code>TRUE</code>.</p></dd>


<dt id="arg-silent">silent<a class="anchor" aria-label="anchor" href="#arg-silent"></a></dt>
<dd><p>Logical. Whether we print the SS loadings,
proportion of variance and the cumulative proportion of
the selected factors or not. Default is <code>TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A length <code>lag * p^2</code> named vector,
<code>p</code> is the number of scales (factors) selected.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Atchley, W. R., Zhao, J., Fernandes, A. D., &amp; Druke, T. (2005).
Solving the protein sequence metric problem.
Proceedings of the National Academy of Sciences of the United States of America,
102(18), 6395-6400.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Nan Xiao &lt;<a href="https://nanx.me" class="external-link">https://nanx.me</a>&gt;</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"protseq/P00750.fasta"</span>, package <span class="op">=</span> <span class="st">"protr"</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">AATopo</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">tprops</span> <span class="op">&lt;-</span> <span class="va">AATopo</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">37</span><span class="op">:</span><span class="fl">41</span>, <span class="fl">43</span><span class="op">:</span><span class="fl">47</span><span class="op">)</span><span class="op">]</span> <span class="co"># select a set of topological descriptors</span></span></span>
<span class="r-in"><span><span class="va">fa</span> <span class="op">&lt;-</span> <span class="fu">extractFAScales</span><span class="op">(</span><span class="va">x</span>, propmat <span class="op">=</span> <span class="va">tprops</span>, factors <span class="op">=</span> <span class="fl">5</span>, lag <span class="op">=</span> <span class="fl">7</span>, silent <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Summary of the factor analysis result: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Call:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> factanal(x = propmat, factors = factors, scores = scores)</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Uniquenesses:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> WhetZ Whetm Whetv Whete Whetp JhetZ Jhetm Jhetv Jhete Jhetp </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> 0.005 0.005 0.005 0.005 0.005 0.005 0.005 0.005 0.005 0.005 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Loadings:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>       Factor1 Factor2 Factor3 Factor4 Factor5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> WhetZ  0.982  -0.166                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Whetm  0.982  -0.166                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Whetv  0.989  -0.133                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Whete  0.982  -0.159                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Whetp  0.989  -0.131                         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> JhetZ -0.399   0.881   0.162   0.192         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Jhetm -0.399   0.881   0.162   0.192         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Jhetv          0.985          -0.131         </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Jhete -0.378   0.872   0.300                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Jhetp          0.970  -0.210                 </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                Factor1 Factor2 Factor3 Factor4 Factor5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> SS loadings      5.313   4.340   0.212   0.107   0.003</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Proportion Var   0.531   0.434   0.021   0.011   0.000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative Var   0.531   0.965   0.986   0.997   0.997</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Test of the hypothesis that 5 factors are sufficient.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The chi square statistic is 472.2 on 5 degrees of freedom.</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> The p-value is 7.96e-100 </span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nan Xiao, Qing-Song Xu, Dong-Sheng Cao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

