<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Scales-Based Descriptors with 20+ classes of Molecular Descriptors — extractDescScales • protr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Scales-Based Descriptors with 20+ classes of Molecular Descriptors — extractDescScales"><meta name="description" content="This function calculates the scales-based descriptors with
molecular descriptors sets calculated by Dragon, Discovery Studio and MOE.
Users can specify which molecular descriptors to select from one of these
deseriptor sets by specify the numerical or character index of the
molecular descriptors in the descriptor set."><meta property="og:description" content="This function calculates the scales-based descriptors with
molecular descriptors sets calculated by Dragon, Discovery Studio and MOE.
Users can specify which molecular descriptors to select from one of these
deseriptor sets by specify the numerical or character index of the
molecular descriptors in the descriptor set."><meta property="og:image" content="https://nanx.me/protr/logo.png"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">protr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7-4</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/protr.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/nanxstats/protr/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Scales-Based Descriptors with 20+ classes of Molecular Descriptors</h1>
      <small class="dont-index">Source: <a href="https://github.com/nanxstats/protr/blob/master/R/pcm-02-extractDescScales.R" class="external-link"><code>R/pcm-02-extractDescScales.R</code></a></small>
      <div class="d-none name"><code>extractDescScales.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function calculates the scales-based descriptors with
molecular descriptors sets calculated by Dragon, Discovery Studio and MOE.
Users can specify which molecular descriptors to select from one of these
deseriptor sets by specify the numerical or character index of the
molecular descriptors in the descriptor set.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">extractDescScales</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">propmat</span>,</span>
<span>  index <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">pc</span>,</span>
<span>  <span class="va">lag</span>,</span>
<span>  scale <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  silent <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>A character vector, as the input protein sequence.</p></dd>


<dt id="arg-propmat">propmat<a class="anchor" aria-label="anchor" href="#arg-propmat"></a></dt>
<dd><p>The matrix containing the descriptor set for the amino acids,
which can be chosen from
<code>AAMOE2D</code>, <code>AAMOE3D</code>, <code>AACPSA</code>, <code>AADescAll</code>,
<code>AA2DACOR</code>, <code>AA3DMoRSE</code>, <code>AAACF</code>, <code>AABurden</code>,
<code>AAConn</code>, <code>AAConst</code>, <code>AAEdgeAdj</code>, <code>AAEigIdx</code>,
<code>AAFGC</code>, <code>AAGeom</code>, <code>AAGETAWAY</code>, <code>AAInfo</code>,
<code>AAMolProp</code>, <code>AARandic</code>, <code>AARDF</code>, <code>AATopo</code>,
<code>AATopoChg</code>, <code>AAWalk</code>, and <code>AAWHIM</code>.</p></dd>


<dt id="arg-index">index<a class="anchor" aria-label="anchor" href="#arg-index"></a></dt>
<dd><p>Integer vector or character vector. Specify which
molecular descriptors to select from one of these deseriptor
sets by specify the numerical or character index of the molecular
descriptors in the descriptor set.
Default is <code>NULL</code>, which means selecting all the molecular
descriptors in this descriptor set.</p></dd>


<dt id="arg-pc">pc<a class="anchor" aria-label="anchor" href="#arg-pc"></a></dt>
<dd><p>Integer. The maximum dimension of the space which the data
are to be represented in. Must be no greater than the number of
amino acid properties provided.</p></dd>


<dt id="arg-lag">lag<a class="anchor" aria-label="anchor" href="#arg-lag"></a></dt>
<dd><p>The lag parameter. Must be less than the amino acids.</p></dd>


<dt id="arg-scale">scale<a class="anchor" aria-label="anchor" href="#arg-scale"></a></dt>
<dd><p>Logical. Should we auto-scale the property matrix
(<code>propmat</code>) before doing MDS? Default is <code>TRUE</code>.</p></dd>


<dt id="arg-silent">silent<a class="anchor" aria-label="anchor" href="#arg-silent"></a></dt>
<dd><p>Logical. Whether we print the standard deviation,
proportion of variance and the cumulative proportion of
the selected principal components or not. Default is <code>TRUE</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A length <code>lag * p^2</code> named vector,
<code>p</code> is the number of scales selected.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Nan Xiao &lt;<a href="https://nanx.me" class="external-link">https://nanx.me</a>&gt;</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"protseq/P00750.fasta"</span>, package <span class="op">=</span> <span class="st">"protr"</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span></span>
<span class="r-in"><span><span class="va">descscales</span> <span class="op">&lt;-</span> <span class="fu">extractDescScales</span><span class="op">(</span></span></span>
<span class="r-in"><span>  <span class="va">x</span>,</span></span>
<span class="r-in"><span>  propmat <span class="op">=</span> <span class="st">"AATopo"</span>, index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">37</span><span class="op">:</span><span class="fl">41</span>, <span class="fl">43</span><span class="op">:</span><span class="fl">47</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>  pc <span class="op">=</span> <span class="fl">5</span>, lag <span class="op">=</span> <span class="fl">7</span>, silent <span class="op">=</span> <span class="cn">FALSE</span></span></span>
<span class="r-in"><span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Summary of the first 5 principal components: </span>
<span class="r-out co"><span class="r-pr">#&gt;</span>                             PC1      PC2       PC3       PC4        PC5</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Standard deviation     2.581537 1.754133 0.4621854 0.1918666 0.08972087</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Proportion of Variance 0.666430 0.307700 0.0213600 0.0036800 0.00080000</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Cumulative Proportion  0.666430 0.974130 0.9954900 0.9991700 0.99998000</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nan Xiao, Qing-Song Xu, Dong-Sheng Cao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.0.</p>
</div>

    </footer></div>





  </body></html>

