<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Function reference â€¢ protr</title><!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png"><link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png"><link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png"><link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png"><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><meta property="og:image" content="https://nanx.me/protr/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light" data-bs-theme="light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">protr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7-1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../articles/protr.html">Get started</a>
</li>
<li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nanxstats/protr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-index">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Function reference</h1>
    </div>

    <div class="section level2">
      <h2 id="protein-and-peptide-sequence-descriptors">Protein and peptide sequence descriptors<a class="anchor" aria-label="anchor" href="#protein-and-peptide-sequence-descriptors"></a></h2>
      
      <div class="section-desc"><p>Functions for calculating protein/peptide sequence descriptors.</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="extractAAC.html">extractAAC()</a></code> 
        </dt>
        <dd>Amino Acid Composition Descriptor</dd>
      </dl><dl><dt>
          
          <code><a href="extractDC.html">extractDC()</a></code> 
        </dt>
        <dd>Dipeptide Composition Descriptor</dd>
      </dl><dl><dt>
          
          <code><a href="extractTC.html">extractTC()</a></code> 
        </dt>
        <dd>Tripeptide Composition Descriptor</dd>
      </dl><dl><dt>
          
          <code><a href="extractMoreauBroto.html">extractMoreauBroto()</a></code> 
        </dt>
        <dd>Normalized Moreau-Broto Autocorrelation Descriptor</dd>
      </dl><dl><dt>
          
          <code><a href="extractMoran.html">extractMoran()</a></code> 
        </dt>
        <dd>Moran Autocorrelation Descriptor</dd>
      </dl><dl><dt>
          
          <code><a href="extractGeary.html">extractGeary()</a></code> 
        </dt>
        <dd>Geary Autocorrelation Descriptor</dd>
      </dl><dl><dt>
          
          <code><a href="extractCTDC.html">extractCTDC()</a></code> 
        </dt>
        <dd>CTD Descriptors - Composition</dd>
      </dl><dl><dt>
          
          <code><a href="extractCTDCClass.html">extractCTDCClass()</a></code> 
        </dt>
        <dd>CTD Descriptors - Composition (with customized amino acid classification support)</dd>
      </dl><dl><dt>
          
          <code><a href="extractCTDDClass.html">extractCTDDClass()</a></code> 
        </dt>
        <dd>CTD Descriptors - Distribution (with customized amino acid classification support)</dd>
      </dl><dl><dt>
          
          <code><a href="extractCTDTClass.html">extractCTDTClass()</a></code> 
        </dt>
        <dd>CTD Descriptors - Transition (with customized amino acid classification support)</dd>
      </dl><dl><dt>
          
          <code><a href="extractCTDT.html">extractCTDT()</a></code> 
        </dt>
        <dd>CTD Descriptors - Transition</dd>
      </dl><dl><dt>
          
          <code><a href="extractCTDD.html">extractCTDD()</a></code> 
        </dt>
        <dd>CTD Descriptors - Distribution</dd>
      </dl><dl><dt>
          
          <code><a href="extractCTriad.html">extractCTriad()</a></code> 
        </dt>
        <dd>Conjoint Triad Descriptor</dd>
      </dl><dl><dt>
          
          <code><a href="extractCTriadClass.html">extractCTriadClass()</a></code> 
        </dt>
        <dd>Conjoint Triad Descriptor (with customized amino acid classification support)</dd>
      </dl><dl><dt>
          
          <code><a href="extractSOCN.html">extractSOCN()</a></code> 
        </dt>
        <dd>Sequence-Order-Coupling Numbers</dd>
      </dl><dl><dt>
          
          <code><a href="extractQSO.html">extractQSO()</a></code> 
        </dt>
        <dd>Quasi-Sequence-Order (QSO) Descriptor</dd>
      </dl><dl><dt>
          
          <code><a href="extractPAAC.html">extractPAAC()</a></code> 
        </dt>
        <dd>Pseudo Amino Acid Composition (PseAAC) Descriptor</dd>
      </dl><dl><dt>
          
          <code><a href="extractAPAAC.html">extractAPAAC()</a></code> 
        </dt>
        <dd>Amphiphilic Pseudo Amino Acid Composition (APseAAC) Descriptor</dd>
      </dl></div><div class="section level2">
      <h2 id="profile-based-protein-and-peptide-sequence-descriptors">Profile-based protein and peptide sequence descriptors<a class="anchor" aria-label="anchor" href="#profile-based-protein-and-peptide-sequence-descriptors"></a></h2>
      
      <div class="section-desc"><p>Functions for calculating the profile-based (PSSM) descriptors.</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="extractPSSM.html">extractPSSM()</a></code> 
        </dt>
        <dd>Compute PSSM (Position-Specific Scoring Matrix) for given protein sequence</dd>
      </dl><dl><dt>
          
          <code><a href="extractPSSMAcc.html">extractPSSMAcc()</a></code> 
        </dt>
        <dd>Profile-based protein representation derived by PSSM (Position-Specific Scoring Matrix) and auto cross covariance</dd>
      </dl><dl><dt>
          
          <code><a href="extractPSSMFeature.html">extractPSSMFeature()</a></code> 
        </dt>
        <dd>Profile-based protein representation derived by PSSM (Position-Specific Scoring Matrix)</dd>
      </dl><dl><dt>
          
          <code><a href="ACC.html">acc()</a></code> 
        </dt>
        <dd>Auto Cross Covariance (ACC) for Generating Scales-Based Descriptors of the Same Length</dd>
      </dl></div><div class="section level2">
      <h2 id="proteochemometric-modeling-pcm-descriptors">Proteochemometric Modeling (PCM) descriptors<a class="anchor" aria-label="anchor" href="#proteochemometric-modeling-pcm-descriptors"></a></h2>
      
      <div class="section-desc"><p>Functions for calculating proteochemometric modeling (PCM) descriptors.</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="extractScales.html">extractScales()</a></code> 
        </dt>
        <dd>Scales-Based Descriptors derived by Principal Components Analysis</dd>
      </dl><dl><dt>
          
          <code><a href="extractScalesGap.html">extractScalesGap()</a></code> 
        </dt>
        <dd>Scales-Based Descriptors derived by Principal Components Analysis (with Gap Support)</dd>
      </dl><dl><dt>
          
          <code><a href="extractProtFP.html">extractProtFP()</a></code> 
        </dt>
        <dd>Amino Acid Properties Based Scales Descriptors (Protein Fingerprint)</dd>
      </dl><dl><dt>
          
          <code><a href="extractProtFPGap.html">extractProtFPGap()</a></code> 
        </dt>
        <dd>Amino Acid Properties Based Scales Descriptors (Protein Fingerprint) with Gap Support</dd>
      </dl><dl><dt>
          
          <code><a href="extractDescScales.html">extractDescScales()</a></code> 
        </dt>
        <dd>Scales-Based Descriptors with 20+ classes of Molecular Descriptors</dd>
      </dl><dl><dt>
          
          <code><a href="extractFAScales.html">extractFAScales()</a></code> 
        </dt>
        <dd>Scales-Based Descriptors derived by Factor Analysis</dd>
      </dl><dl><dt>
          
          <code><a href="extractMDSScales.html">extractMDSScales()</a></code> 
        </dt>
        <dd>Scales-Based Descriptors derived by Multidimensional Scaling</dd>
      </dl><dl><dt>
          
          <code><a href="extractBLOSUM.html">extractBLOSUM()</a></code> 
        </dt>
        <dd>BLOSUM and PAM Matrix-Derived Descriptors</dd>
      </dl></div><div class="section level2">
      <h2 id="similarity-measures-between-proteins">Similarity measures between proteins<a class="anchor" aria-label="anchor" href="#similarity-measures-between-proteins"></a></h2>
      
      <div class="section-desc"><p>Functions for calculating protein sequence alignment based similarity measures and GO-based semantic similarity measures.</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="parSeqSim.html">parSeqSim()</a></code> 
        </dt>
        <dd>Parallel Protein Sequence Similarity Calculation Based on Sequence Alignment (In-Memory Version)</dd>
      </dl><dl><dt>
          
          <code><a href="parSeqSimDisk.html">parSeqSimDisk()</a></code> 
        </dt>
        <dd>Parallel Protein Sequence Similarity Calculation Based on Sequence Alignment (Disk-Based Version)</dd>
      </dl><dl><dt>
          
          <code><a href="crossSetSim.html">crossSetSim()</a></code> 
        </dt>
        <dd>Parallel Protein Sequence Similarity Calculation Between Two Sets Based on Sequence Alignment (In-Memory Version)</dd>
      </dl><dl><dt>
          
          <code><a href="crossSetSimDisk.html">crossSetSimDisk()</a></code> 
        </dt>
        <dd>Parallel Protein Sequence Similarity Calculation Between Two Sets Based on Sequence Alignment (Disk-Based Version)</dd>
      </dl><dl><dt>
          
          <code><a href="twoSeqSim.html">twoSeqSim()</a></code> 
        </dt>
        <dd>Protein Sequence Alignment for Two Protein Sequences</dd>
      </dl><dl><dt>
          
          <code><a href="parGOSim.html">parGOSim()</a></code> 
        </dt>
        <dd>Protein Similarity Calculation based on Gene Ontology (GO) Similarity</dd>
      </dl><dl><dt>
          
          <code><a href="twoGOSim.html">twoGOSim()</a></code> 
        </dt>
        <dd>Protein Similarity Calculation based on Gene Ontology (GO) Similarity</dd>
      </dl></div><div class="section level2">
      <h2 id="pre-process-protein-sequences">Pre-process protein sequences<a class="anchor" aria-label="anchor" href="#pre-process-protein-sequences"></a></h2>
      
      <div class="section-desc"><p>Helper Functions for pre-processing protein sequences.</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="getUniProt.html">getUniProt()</a></code> 
        </dt>
        <dd>Retrieve Protein Sequences from UniProt by Protein ID</dd>
      </dl><dl><dt>
          
          <code><a href="readFASTA.html">readFASTA()</a></code> 
        </dt>
        <dd>Read Protein Sequences in FASTA Format</dd>
      </dl><dl><dt>
          
          <code><a href="readPDB.html">readPDB()</a></code> 
        </dt>
        <dd>Read Protein Sequences in PDB Format</dd>
      </dl><dl><dt>
          
          <code><a href="protcheck.html">protcheck()</a></code> 
        </dt>
        <dd>Protein sequence amino acid type sanity check</dd>
      </dl><dl><dt>
          
          <code><a href="protseg.html">protseg()</a></code> 
        </dt>
        <dd>Protein Sequence Segmentation/Partition</dd>
      </dl><dl><dt>
          
          <code><a href="removeGaps.html">removeGaps()</a></code> 
        </dt>
        <dd>Remove or replace gaps from protein sequences.</dd>
      </dl><dl><dt>
          
          <code><a href="protr-package.html">protr</a></code> <code><a href="protr-package.html">protr-package</a></code> 
        </dt>
        <dd>protr: Generating Various Numerical Representation Schemes for Protein Sequences</dd>
      </dl></div><div class="section level2">
      <h2 id="precomputed-molecular-descriptors-for-the-amino-acids">Precomputed molecular descriptors for the 20 amino acids<a class="anchor" aria-label="anchor" href="#precomputed-molecular-descriptors-for-the-amino-acids"></a></h2>
      
      <div class="section-desc"><p>Precomputed molecular descriptors for the 20 amino acids.</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="AAindex.html">AAindex</a></code> 
        </dt>
        <dd>AAindex Data of 544 Physicochemical and Biological Properties for 20 Amino Acids</dd>
      </dl><dl><dt>
          
          <code><a href="AAMetaInfo.html">AAMetaInfo</a></code> 
        </dt>
        <dd>Meta Information for the 20 Amino Acids</dd>
      </dl><dl><dt>
          
          <code><a href="OptAA3d.html">OptAA3d</a></code> 
        </dt>
        <dd>OptAA3d.sdf - 20 Amino Acids Optimized with MOE 2011.10 (Semiempirical AM1)</dd>
      </dl><dl><dt>
          
          <code><a href="AA2DACOR.html">AA2DACOR</a></code> 
        </dt>
        <dd>2D Autocorrelations Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AA3DMoRSE.html">AA3DMoRSE</a></code> 
        </dt>
        <dd>3D-MoRSE Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAACF.html">AAACF</a></code> 
        </dt>
        <dd>Atom-Centred Fragments Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AABurden.html">AABurden</a></code> 
        </dt>
        <dd>Burden Eigenvalues Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAConn.html">AAConn</a></code> 
        </dt>
        <dd>Connectivity Indices Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAConst.html">AAConst</a></code> 
        </dt>
        <dd>Constitutional Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AACPSA.html">AACPSA</a></code> 
        </dt>
        <dd>CPSA Descriptors for 20 Amino Acids calculated by Discovery Studio</dd>
      </dl><dl><dt>
          
          <code><a href="AADescAll.html">AADescAll</a></code> 
        </dt>
        <dd>All 2D Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAEdgeAdj.html">AAEdgeAdj</a></code> 
        </dt>
        <dd>Edge Adjacency Indices Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAEigIdx.html">AAEigIdx</a></code> 
        </dt>
        <dd>Eigenvalue-Based Indices Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAFGC.html">AAFGC</a></code> 
        </dt>
        <dd>Functional Group Counts Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAGeom.html">AAGeom</a></code> 
        </dt>
        <dd>Geometrical Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAGETAWAY.html">AAGETAWAY</a></code> 
        </dt>
        <dd>GETAWAY Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAInfo.html">AAInfo</a></code> 
        </dt>
        <dd>Information Indices Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAMOE2D.html">AAMOE2D</a></code> 
        </dt>
        <dd>2D Descriptors for 20 Amino Acids calculated by MOE 2011.10</dd>
      </dl><dl><dt>
          
          <code><a href="AAMOE3D.html">AAMOE3D</a></code> 
        </dt>
        <dd>3D Descriptors for 20 Amino Acids calculated by MOE 2011.10</dd>
      </dl><dl><dt>
          
          <code><a href="AAMolProp.html">AAMolProp</a></code> 
        </dt>
        <dd>Molecular Properties Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AARandic.html">AARandic</a></code> 
        </dt>
        <dd>Randic Molecular Profiles Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AARDF.html">AARDF</a></code> 
        </dt>
        <dd>RDF Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AATopo.html">AATopo</a></code> 
        </dt>
        <dd>Topological Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AATopoChg.html">AATopoChg</a></code> 
        </dt>
        <dd>Topological Charge Indices Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAWalk.html">AAWalk</a></code> 
        </dt>
        <dd>Walk and Path Counts Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl><dl><dt>
          
          <code><a href="AAWHIM.html">AAWHIM</a></code> 
        </dt>
        <dd>WHIM Descriptors for 20 Amino Acids calculated by Dragon</dd>
      </dl></div><div class="section level2">
      <h2 id="blosum-and-pam-matrices-for-the-amino-acids">BLOSUM and PAM matrices for the 20 amino acids<a class="anchor" aria-label="anchor" href="#blosum-and-pam-matrices-for-the-amino-acids"></a></h2>
      
      <div class="section-desc"><p>BLOSUM and PAM matrices for the 20 amino acids.</p></div>

      
    </div><div id="" class="section level2">
      
      
      

      <dl><dt>
          
          <code><a href="AABLOSUM45.html">AABLOSUM45</a></code> 
        </dt>
        <dd>BLOSUM45 Matrix for 20 Amino Acids</dd>
      </dl><dl><dt>
          
          <code><a href="AABLOSUM50.html">AABLOSUM50</a></code> 
        </dt>
        <dd>BLOSUM50 Matrix for 20 Amino Acids</dd>
      </dl><dl><dt>
          
          <code><a href="AABLOSUM62.html">AABLOSUM62</a></code> 
        </dt>
        <dd>BLOSUM62 Matrix for 20 Amino Acids</dd>
      </dl><dl><dt>
          
          <code><a href="AABLOSUM80.html">AABLOSUM80</a></code> 
        </dt>
        <dd>BLOSUM80 Matrix for 20 Amino Acids</dd>
      </dl><dl><dt>
          
          <code><a href="AABLOSUM100.html">AABLOSUM100</a></code> 
        </dt>
        <dd>BLOSUM100 Matrix for 20 Amino Acids</dd>
      </dl><dl><dt>
          
          <code><a href="AAPAM30.html">AAPAM30</a></code> 
        </dt>
        <dd>PAM30 Matrix for 20 Amino Acids</dd>
      </dl><dl><dt>
          
          <code><a href="AAPAM40.html">AAPAM40</a></code> 
        </dt>
        <dd>PAM40 Matrix for 20 Amino Acids</dd>
      </dl><dl><dt>
          
          <code><a href="AAPAM70.html">AAPAM70</a></code> 
        </dt>
        <dd>PAM70 Matrix for 20 Amino Acids</dd>
      </dl><dl><dt>
          
          <code><a href="AAPAM120.html">AAPAM120</a></code> 
        </dt>
        <dd>PAM120 Matrix for 20 Amino Acids</dd>
      </dl><dl><dt>
          
          <code><a href="AAPAM250.html">AAPAM250</a></code> 
        </dt>
        <dd>PAM250 Matrix for 20 Amino Acids</dd>
      </dl></div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Nan Xiao, Qing-Song Xu, Dong-Sheng Cao.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

