<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="protr">
<title>protr: R package for generating various numerical representation schemes of protein sequences â€¢ protr</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="protr: R package for generating various numerical representation schemes of protein sequences">
<meta property="og:description" content="protr">
<meta property="og:image" content="https://nanx.me/protr/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">protr</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.6-3</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/protr.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/nanxstats/protr/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>protr: R package for generating various numerical representation schemes of protein sequences</h1>
                        <h4 data-toc-skip class="author">Nan Xiao &lt;<a href="https://nanx.me" class="external-link uri">https://nanx.me</a>&gt;</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/nanxstats/protr/blob/HEAD/vignettes/protr.Rmd" class="external-link"><code>vignettes/protr.Rmd</code></a></small>
      <div class="d-none name"><code>protr.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>The protr package offers a unique and comprehensive toolkit for
generating various numerical representation schemes of protein
sequences. The descriptors included are extensively utilized in
bioinformatics and chemogenomics research.</p>
<p>The commonly used descriptors listed in protr include amino acid
composition, autocorrelation, CTD, conjoint traid, quasi-sequence order,
pseudo amino acid composition, and profile-based descriptors derived by
Position-Specific Scoring Matrix (PSSM).</p>
<p>The descriptors for proteochemometric (PCM) modeling, includes the
scales-based descriptors derived by principal components analysis,
factor analysis, multidimensional scaling, amino acid properties
(AAindex), 20+ classes of 2D and 3D molecular descriptors (Topological,
WHIM, VHSE, etc.), and BLOSUM/PAM matrix-derived descriptors.</p>
<p>The protr package also implemented parallelized similarity
computation derived by pairwise protein sequence alignment and Gene
Ontology (GO) semantic similarity measures. ProtrWeb, the web
application built on protr, can be accessed from <a href="http://protr.org" class="external-link uri">http://protr.org</a>.</p>
<p>If you find protr useful in your research, please feel free to cite
our paper:</p>
<blockquote>
Nan Xiao, Dong-Sheng Cao, Min-Feng Zhu, and Qing-Song Xu. (2015).
protr/ProtrWeb: R package and web server for generating various
numerical representation schemes of protein sequences.
<em>Bioinformatics</em> 31 (11), 1857-1859.
</blockquote>
<p>BibTeX entry:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bibtex"><code class="sourceCode bibtex"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="va">@article</span>{<span class="ot">Xiao2015</span>,</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>  <span class="dt">author</span>  = {Xiao, Nan and Cao, Dong-Sheng and Zhu, Min-Feng and Xu, Qing-Song.},</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>  <span class="dt">title</span>   = {protr/{ProtrWeb}: {R} package and web server for generating various numerical representation schemes of protein sequences},</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>  <span class="dt">journal</span> = {Bioinformatics},</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="dt">year</span>    = {2015},</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="dt">volume</span>  = {31},</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  <span class="dt">number</span>  = {11},</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>  <span class="dt">pages</span>   = {1857--1859},</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>  <span class="dt">doi</span>     = {10.1093/bioinformatics/btv042}</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>}</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="an-example-predictive-modeling-workflow">An example predictive modeling workflow<a class="anchor" aria-label="anchor" href="#an-example-predictive-modeling-workflow"></a>
</h2>
<p>Here we use the subcellular localization dataset of human proteins
presented in <span class="citation">Chou and Shen (2008)</span> to
demonstrate the basic usage of protr.</p>
<p>The complete dataset includes 3,134 protein sequences (2,750
different proteins), classified into 14 human subcellular locations. We
selected two classes of proteins as our benchmark dataset. Class 1
contains 325 <em>extracell</em> proteins, and class 2 includes 307
<em>mitochondrion</em> proteins. Here we aim to build a random forest
classification model to classify these two types of proteins.</p>
<p>First, we load the protr package, then read the protein sequences
stored in two separated FASTA files with <code><a href="../reference/readFASTA.html">readFASTA()</a></code>:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://nanx.me/protr/">"protr"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co"># load FASTA files</span></span>
<span><span class="co"># (system.file is for accessing example file in protr package, </span></span>
<span><span class="co"># replace it with your path)</span></span>
<span><span class="va">extracell</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span>  </span>
<span>  <span class="st">"protseq/extracell.fasta"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"protr"</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">mitonchon</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"protseq/mitochondrion.fasta"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"protr"</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>To read protein sequences stored in PDB format files, use
<code><a href="../reference/readPDB.html">readPDB()</a></code> instead. The loaded sequences will be stored as
two lists in R, and each component in the list is a character string
representing one protein sequence. In this case, there are 325
<em>extracell</em> protein sequences and 306 <em>mitonchon</em> protein
sequences:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">extracell</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [1] 325</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mitonchon</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [1] 306</span></span></code></pre></div>
<p>To ensure that the protein sequences only have the 20 standard amino
acid types which is usually required for the descriptor computation, we
use the <code><a href="../reference/protcheck.html">protcheck()</a></code> function to do the amino acid type
sanity check and remove the <em>non-standard</em> sequences:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">extracell</span> <span class="op">&lt;-</span> <span class="va">extracell</span><span class="op">[</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">extracell</span>, <span class="va">protcheck</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">mitonchon</span> <span class="op">&lt;-</span> <span class="va">mitonchon</span><span class="op">[</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">mitonchon</span>, <span class="va">protcheck</span><span class="op">)</span><span class="op">)</span><span class="op">]</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">extracell</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [1] 323</span></span></code></pre></div>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mitonchon</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [1] 304</span></span></code></pre></div>
<p>Two protein sequences were removed from each class. For the remaining
sequences, we calculate the Type II PseAAC descriptor, i.e., the
amphiphilic pseudo amino acid composition (APseAAC) descriptor <span class="citation">(Chou, 2005)</span> and make class labels for
classification modeling.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># calculate APseAAC descriptors</span></span>
<span><span class="va">x1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">extracell</span>, <span class="va">extractAPAAC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">mitonchon</span>, <span class="va">extractAPAAC</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">x1</span>, <span class="va">x2</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># make class labels</span></span>
<span><span class="va">labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">extracell</span><span class="op">)</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">mitonchon</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>In protr, the functions of commonly used descriptors for protein
sequences and proteochemometric (PCM) modeling descriptors are named
after <code>extract...()</code>.</p>
<p>Next, we will split the data into a 75% training set and a 25% test
set.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1001</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># split training and test set</span></span>
<span><span class="va">tr.idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x1</span><span class="op">)</span> <span class="op">*</span> <span class="fl">0.75</span><span class="op">)</span><span class="op">)</span>,</span>
<span>  <span class="fu"><a href="https://rdrr.io/pkg/AnnotationDbi/man/Bimap-envirAPI.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x1</span><span class="op">)</span> <span class="op">+</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x2</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x2</span><span class="op">)</span> <span class="op">*</span> <span class="fl">0.75</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">te.idx</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span>, <span class="va">tr.idx</span><span class="op">)</span></span>
<span></span>
<span><span class="va">x.tr</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="va">tr.idx</span>, <span class="op">]</span></span>
<span><span class="va">x.te</span> <span class="op">&lt;-</span> <span class="va">x</span><span class="op">[</span><span class="va">te.idx</span>, <span class="op">]</span></span>
<span><span class="va">y.tr</span> <span class="op">&lt;-</span> <span class="va">labels</span><span class="op">[</span><span class="va">tr.idx</span><span class="op">]</span></span>
<span><span class="va">y.te</span> <span class="op">&lt;-</span> <span class="va">labels</span><span class="op">[</span><span class="va">te.idx</span><span class="op">]</span></span></code></pre></div>
<p>We will train a random forest classification model on the training
set with 5-fold cross-validation, using the <code>randomForest</code>
package.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://www.stat.berkeley.edu/~breiman/RandomForests/" class="external-link">"randomForest"</a></span><span class="op">)</span></span>
<span><span class="va">rf.fit</span> <span class="op">&lt;-</span> <span class="fu">randomForest</span><span class="op">(</span><span class="va">x.tr</span>, <span class="va">y.tr</span>, cv.fold <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="va">rf.fit</span></span></code></pre></div>
<p>The training result is:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Call:</span></span>
<span><span class="co">##  randomForest(x = x.tr, y = y.tr, cv.fold = 5)</span></span>
<span><span class="co">##                Type of random forest: classification</span></span>
<span><span class="co">##                      Number of trees: 500</span></span>
<span><span class="co">## No. of variables tried at each split: 8</span></span>
<span><span class="co">##</span></span>
<span><span class="co">##         OOB estimate of  error rate: 25.11%</span></span>
<span><span class="co">## Confusion matrix:</span></span>
<span><span class="co">##     0   1 class.error</span></span>
<span><span class="co">## 0 196  46   0.1900826</span></span>
<span><span class="co">## 1  72 156   0.3157895</span></span></code></pre></div>
<p>With the model trained on the training set, we predict on the test
set and plot the ROC curve with the <code>pROC</code> package, as is
shown in Figure 1.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># predict on test set</span></span>
<span><span class="va">rf.pred</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html" class="external-link">predict</a></span><span class="op">(</span><span class="va">rf.fit</span>, newdata <span class="op">=</span> <span class="va">x.te</span>, type <span class="op">=</span> <span class="st">"prob"</span><span class="op">)</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># plot ROC curve</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="http://expasy.org/tools/pROC/" class="external-link">"pROC"</a></span><span class="op">)</span></span>
<span><span class="fu">plot.roc</span><span class="op">(</span><span class="va">y.te</span>, <span class="va">rf.pred</span>, grid <span class="op">=</span> <span class="cn">TRUE</span>, print.auc <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>The area under the ROC curve (AUC) is:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Call:</span></span>
<span><span class="co">## plot.roc.default(x = y.te, predictor = rf.pred, col = "#0080ff",</span></span>
<span><span class="co">##                  grid = TRUE, print.auc = TRUE)</span></span>
<span><span class="co">##</span></span>
<span><span class="co">## Data: rf.pred in 81 controls (y.te 0) &gt; 76 cases (y.te 1).</span></span>
<span><span class="co">## Area under the curve: 0.8697</span></span></code></pre></div>
<img src="figures/roc.png" width="60%" class="img-responsive" align="center" alt="ROC curve for the test set of protein subcellular localization data"><p align="center">
Figure 1: ROC curve for the test set of protein subcellular localization
data
</p>
</div>
<div class="section level2">
<h2 id="package-overview">Package overview<a class="anchor" aria-label="anchor" href="#package-overview"></a>
</h2>
<p>The protr package <span class="citation">(Xiao <em>et al.</em>,
2015)</span> implemented most of the state-of-the-art protein sequence
descriptors with R. Generally, each type of the descriptors (features)
can be calculated with a function named <code>extractX()</code> in the
protr package, where <code>X</code> stands for the abbrevation of the
descriptor name. The descriptors are:</p>
<ul>
<li>Amino acid composition
<ul>
<li>
<code><a href="../reference/extractAAC.html">extractAAC()</a></code> - Amino acid composition</li>
<li>
<code><a href="../reference/extractDC.html">extractDC()</a></code> - Dipeptide composition</li>
<li>
<code><a href="../reference/extractTC.html">extractTC()</a></code> - Tripeptide composition</li>
</ul>
</li>
<li>Autocorrelation
<ul>
<li>
<code><a href="../reference/extractMoreauBroto.html">extractMoreauBroto()</a></code> - Normalized Moreau-Broto
autocorrelation</li>
<li>
<code><a href="../reference/extractMoran.html">extractMoran()</a></code> - Moran autocorrelation</li>
<li>
<code><a href="../reference/extractGeary.html">extractGeary()</a></code> - Geary autocorrelation</li>
</ul>
</li>
<li>CTD descriptors
<ul>
<li>
<code><a href="../reference/extractCTDC.html">extractCTDC()</a></code> - Composition</li>
<li>
<code><a href="../reference/extractCTDT.html">extractCTDT()</a></code> - Transition</li>
<li>
<code><a href="../reference/extractCTDD.html">extractCTDD()</a></code> - Distribution</li>
</ul>
</li>
<li>Conjoint triad descriptors
<ul>
<li>
<code><a href="../reference/extractCTriad.html">extractCTriad()</a></code> - Conjoint triad descriptors</li>
</ul>
</li>
<li>Quasi-sequence-order descriptors
<ul>
<li>
<code><a href="../reference/extractSOCN.html">extractSOCN()</a></code> - Sequence-order-coupling number</li>
<li>
<code><a href="../reference/extractQSO.html">extractQSO()</a></code> - Quasi-sequence-order descriptors</li>
</ul>
</li>
<li>Pseudo-amino acid composition
<ul>
<li>
<code><a href="../reference/extractPAAC.html">extractPAAC()</a></code> - Pseudo-amino acid composition
(PseAAC)</li>
<li>
<code><a href="../reference/extractAPAAC.html">extractAPAAC()</a></code> - Amphiphilic pseudo-amino acid
composition (APseAAC)</li>
</ul>
</li>
<li>Profile-based descriptors
<ul>
<li><code><a href="../reference/extractPSSM.html">extractPSSM()</a></code></li>
<li><code><a href="../reference/extractPSSMAcc.html">extractPSSMAcc()</a></code></li>
<li><code><a href="../reference/extractPSSMFeature.html">extractPSSMFeature()</a></code></li>
</ul>
</li>
</ul>
<p>The descriptors commonly used in Proteochemometric Modeling (PCM)
implemented in protr include:</p>
<ul>
<li>
<code><a href="../reference/extractScales.html">extractScales()</a></code>, <code><a href="../reference/extractScalesGap.html">extractScalesGap()</a></code> -
Scales-based descriptors derived by Principal Components Analysis
<ul>
<li>
<code><a href="../reference/extractProtFP.html">extractProtFP()</a></code>, <code><a href="../reference/extractProtFPGap.html">extractProtFPGap()</a></code> -
Scales-based descriptors derived by amino acid properties from AAindex
(a.k.a. <em>Protein Fingerprint</em>)</li>
<li>
<code><a href="../reference/extractDescScales.html">extractDescScales()</a></code> - Scales-based descriptors derived
by 20+ classes of 2D and 3D molecular descriptors (Topological, WHIM,
VHSE, etc.)</li>
</ul>
</li>
<li>
<code><a href="../reference/extractFAScales.html">extractFAScales()</a></code> - Scales-based descriptors derived by
Factor Analysis</li>
<li>
<code><a href="../reference/extractMDSScales.html">extractMDSScales()</a></code> - Scales-based descriptors derived
by Multidimensional Scaling</li>
<li>
<code><a href="../reference/extractBLOSUM.html">extractBLOSUM()</a></code> - BLOSUM and PAM matrix-derived
descriptors</li>
</ul>
<p>The protr package integrates the function of parallelized similarity
score computation derived by local or global protein sequence alignment
between a list of protein sequences, the sequence alignment computation
is provided by <code>Biostrings</code>, the corresponding functions
listed in the <code>protr</code> package include:</p>
<ul>
<li>
<code><a href="../reference/twoSeqSim.html">twoSeqSim()</a></code> - Similarity calculation derived by
sequence alignment between two protein sequences</li>
<li>
<code><a href="../reference/parSeqSim.html">parSeqSim()</a></code> - Parallelized pairwise similarity
calculation with a list of protein sequences (in-memory version)</li>
<li>
<code><a href="../reference/parSeqSimDisk.html">parSeqSimDisk()</a></code> - Parallelized pairwise similarity
calculation with a list of protein sequences (disk-based version)</li>
</ul>
<p>The protr package also integrates the function of parallelized
similarity score computation derived by Gene Ontology (GO) semantic
similarity measures between a list of GO terms / Entrez Gene IDs, the GO
similarity computation is provided by <code>GOSemSim</code>, the
corresponding functions listed in the protr package include:</p>
<ul>
<li>
<code><a href="../reference/twoGOSim.html">twoGOSim()</a></code> - Similarity calculation derived by GO-terms
semantic similarity measures between two GO terms / Entrez Gene
IDs;</li>
<li>
<code><a href="../reference/parGOSim.html">parGOSim()</a></code> - Pairwise similarity calculation with a
list of GO terms / Entrez Gene IDs.</li>
</ul>
<p>To use the <code><a href="../reference/parSeqSim.html">parSeqSim()</a></code> function, we suggest the users to
install the packages <code>foreach</code> and <code>doParallel</code>
first, in order to make the parallelized pairwise similarity computation
available.</p>
<p>In the following sections, we will introduce the descriptors and
function usage in this order.</p>
</div>
<div class="section level2">
<h2 id="commonly-used-descriptors">Commonly used descriptors<a class="anchor" aria-label="anchor" href="#commonly-used-descriptors"></a>
</h2>
<p><strong>Note</strong>: Users need to intelligently evaluate the
underlying details of the descriptors provided, instead of using protr
with their data blindly, especially for the descriptor types with more
flexibility. It would be wise for the users to use some negative and
positive control comparisons where relevant to help guide interpretation
of the results.</p>
<p>A protein or peptide sequence with <span class="math inline">\(N\)</span> amino acid residues can be generally
represented as <span class="math inline">\(\{\,R_1, R_2, \ldots,
R_n\,\}\)</span>, where <span class="math inline">\(R_i\)</span>
represents the residue at the <span class="math inline">\(i\)</span>-th
position in the sequence. The labels <span class="math inline">\(i\)</span> and <span class="math inline">\(j\)</span> are used to index amino acid position
in a sequence, and <span class="math inline">\(r\)</span>, <span class="math inline">\(s\)</span>, <span class="math inline">\(t\)</span>
are used to represent the amino acid type. The computed descriptors are
roughly categorized into 4 groups according to their major
applications.</p>
<p>A protein sequence can be partitioned equally into segments. The
descriptors designed for the complete sequence, can be often applied to
each individual segment.</p>
<div class="section level3">
<h3 id="amino-acid-composition-descriptor">Amino acid composition descriptor<a class="anchor" aria-label="anchor" href="#amino-acid-composition-descriptor"></a>
</h3>
<p>The amino acid composition describes the fraction of each amino acid
type within a protein sequence. The fractions of all 20 natural amino
acids are calculated as:</p>
<p><span class="math display">\[
f(r) = \frac{N_r}{N} \quad r = 1, 2, \ldots, 20.
\]</span> where <span class="math inline">\(N_r\)</span> is the number
of the amino acid type <span class="math inline">\(r\)</span> and <span class="math inline">\(N\)</span> is the length of the sequence.</p>
<p>As was described above, we can use the function
<code><a href="../reference/extractAAC.html">extractAAC()</a></code> to extract the descriptors (features) from
protein sequences:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://nanx.me/protr/">"protr"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"protseq/P00750.fasta"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"protr"</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="fu"><a href="../reference/extractAAC.html">extractAAC</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt;          A          R          N          D          C          E          Q </span></span>
<span><span class="co">#&gt; 0.06405694 0.07117438 0.03914591 0.05160142 0.06761566 0.04804270 0.04804270 </span></span>
<span><span class="co">#&gt;          G          H          I          L          K          M          F </span></span>
<span><span class="co">#&gt; 0.08185053 0.03024911 0.03558719 0.07651246 0.03914591 0.01245552 0.03202847 </span></span>
<span><span class="co">#&gt;          P          S          T          W          Y          V </span></span>
<span><span class="co">#&gt; 0.05338078 0.08896797 0.04448399 0.02313167 0.04270463 0.04982206</span></span></code></pre></div>
<p>Here, with the function <code><a href="../reference/readFASTA.html">readFASTA()</a></code> we loaded a single
protein sequence (P00750, Tissue-type plasminogen activator) from a
FASTA format file. Then extracted the AAC descriptors with
<code><a href="../reference/extractAAC.html">extractAAC()</a></code>. The result returned is a named vector, whose
elements are tagged with the name of each amino acid.</p>
</div>
<div class="section level3">
<h3 id="dipeptide-composition-descriptor">Dipeptide composition descriptor<a class="anchor" aria-label="anchor" href="#dipeptide-composition-descriptor"></a>
</h3>
<p>Dipeptide composition gives a 400-dimensional descriptor, defined
as:</p>
<p><span class="math display">\[
f(r, s) = \frac{N_{rs}}{N - 1} \quad r, s = 1, 2, \ldots, 20.
\]</span></p>
<p>where <span class="math inline">\(N_{rs}\)</span> is the number of
dipeptide represented by amino acid type <span class="math inline">\(r\)</span> and type <span class="math inline">\(s\)</span>. Similar to <code><a href="../reference/extractAAC.html">extractAAC()</a></code>,
here we use <code><a href="../reference/extractDC.html">extractDC()</a></code> to compute the descriptors:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractDC.html">extractDC</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">dc</span>, n <span class="op">=</span> <span class="fl">30L</span><span class="op">)</span></span>
<span><span class="co">#&gt;          AA          RA          NA          DA          CA          EA </span></span>
<span><span class="co">#&gt; 0.003565062 0.003565062 0.000000000 0.007130125 0.003565062 0.003565062 </span></span>
<span><span class="co">#&gt;          QA          GA          HA          IA          LA          KA </span></span>
<span><span class="co">#&gt; 0.007130125 0.007130125 0.001782531 0.003565062 0.001782531 0.001782531 </span></span>
<span><span class="co">#&gt;          MA          FA          PA          SA          TA          WA </span></span>
<span><span class="co">#&gt; 0.000000000 0.005347594 0.003565062 0.007130125 0.003565062 0.000000000 </span></span>
<span><span class="co">#&gt;          YA          VA          AR          RR          NR          DR </span></span>
<span><span class="co">#&gt; 0.000000000 0.000000000 0.003565062 0.007130125 0.005347594 0.001782531 </span></span>
<span><span class="co">#&gt;          CR          ER          QR          GR          HR          IR </span></span>
<span><span class="co">#&gt; 0.005347594 0.005347594 0.000000000 0.007130125 0.001782531 0.003565062</span></span></code></pre></div>
<p>Here we only showed the first 30 elements of the result vector and
omitted the rest of the result. The element names of the returned vector
are self-explanatory as before.</p>
</div>
<div class="section level3">
<h3 id="tripeptide-composition-descriptor">Tripeptide composition descriptor<a class="anchor" aria-label="anchor" href="#tripeptide-composition-descriptor"></a>
</h3>
<p>Tripeptide composition gives a 8000-dimensional descriptor, defined
as:</p>
<p><span class="math display">\[
f(r, s, t) = \frac{N_{rst}}{N - 2} \quad r, s, t = 1, 2, \ldots, 20
\]</span></p>
<p>where <span class="math inline">\(N_{rst}\)</span> is the number of
tripeptides represented by amino acid type <span class="math inline">\(r\)</span>, <span class="math inline">\(s\)</span>, and <span class="math inline">\(t\)</span>. With function
<code><a href="../reference/extractTC.html">extractTC()</a></code>, we can easily obtain the length-8000
descriptor, to save some space, here we also omitted the long
outputs:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractTC.html">extractTC</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">tc</span>, n <span class="op">=</span> <span class="fl">36L</span><span class="op">)</span></span>
<span><span class="co">#&gt;         AAA         RAA         NAA         DAA         CAA         EAA </span></span>
<span><span class="co">#&gt; 0.000000000 0.000000000 0.000000000 0.000000000 0.000000000 0.000000000 </span></span>
<span><span class="co">#&gt;         QAA         GAA         HAA         IAA         LAA         KAA </span></span>
<span><span class="co">#&gt; 0.001785714 0.000000000 0.000000000 0.000000000 0.000000000 0.000000000 </span></span>
<span><span class="co">#&gt;         MAA         FAA         PAA         SAA         TAA         WAA </span></span>
<span><span class="co">#&gt; 0.000000000 0.000000000 0.000000000 0.001785714 0.000000000 0.000000000 </span></span>
<span><span class="co">#&gt;         YAA         VAA         ARA         RRA         NRA         DRA </span></span>
<span><span class="co">#&gt; 0.000000000 0.000000000 0.000000000 0.000000000 0.000000000 0.000000000 </span></span>
<span><span class="co">#&gt;         CRA         ERA         QRA         GRA         HRA         IRA </span></span>
<span><span class="co">#&gt; 0.000000000 0.000000000 0.000000000 0.001785714 0.000000000 0.000000000 </span></span>
<span><span class="co">#&gt;         LRA         KRA         MRA         FRA         PRA         SRA </span></span>
<span><span class="co">#&gt; 0.000000000 0.000000000 0.000000000 0.000000000 0.000000000 0.000000000</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="autocorrelation-descriptors">Autocorrelation descriptors<a class="anchor" aria-label="anchor" href="#autocorrelation-descriptors"></a>
</h3>
<p>Autocorrelation descriptors are defined based on the distribution of
amino acid properties along the sequence. The amino acid properties used
here are various types of amino acids index (Retrieved from the <a href="https://www.genome.jp/dbget/aaindex.html" class="external-link">AAindex database</a>,
see <span class="citation">Kawashima <em>et al.</em> (1999)</span>,
<span class="citation">Kawashima and Kanehisa (2000)</span>, and <span class="citation">Kawashima <em>et al.</em> (2008)</span>; see Figure 2
for an illustrated example). Three types of autocorrelation descriptors
are defined here and described below.</p>
<p>All the amino acid indices are centralized and standardized before
the calculation, i.e.</p>
<p><span class="math display">\[
P_r = \frac{P_r - \bar{P}}{\sigma}
\]</span></p>
<p>where <span class="math inline">\(\bar{P}\)</span> is the average of
the property of the 20 amino acids:</p>
<p><span class="math display">\[
\bar{P} = \frac{\sum_{r=1}^{20} P_r}{20} \quad \textrm{and} \quad \sigma
= \sqrt{\frac{1}{2} \sum_{r=1}^{20} (P_r - \bar{P})^2}
\]</span></p>
<img src="figures/AAindex.png" width="80%" class="img-responsive" align="center" alt="An illustrated example in the AAIndex database"><p align="center">
Figure 2: An illustrated example in the AAIndex database
</p>
<div class="section level4">
<h4 id="normalized-moreau-broto-autocorrelation-descriptors">Normalized Moreau-Broto autocorrelation descriptors<a class="anchor" aria-label="anchor" href="#normalized-moreau-broto-autocorrelation-descriptors"></a>
</h4>
<p>For protein sequences, the Moreau-Broto autocorrelation descriptors
can be defined as:</p>
<p><span class="math display">\[
AC(d) = \sum_{i=1}^{N-d} P_i P_{i + d} \quad d = 1, 2, \ldots,
\textrm{nlag}
\]</span></p>
<p>where <span class="math inline">\(d\)</span> is called the lag of the
autocorrelation; <span class="math inline">\(P_i\)</span> and <span class="math inline">\(P_{i+d}\)</span> are the properties of the amino
acids at position <span class="math inline">\(i\)</span> and <span class="math inline">\(i+d\)</span>; <span class="math inline">\(\textrm{nlag}\)</span> is the maximum value of the
lag.</p>
<p>The normalized Moreau-Broto autocorrelation descriptors are defined
as:</p>
<p><span class="math display">\[
ATS(d) = \frac{AC(d)}{N-d} \quad d = 1, 2, \ldots, \textrm{nlag}
\]</span></p>
<p>The corresponding function for this descriptor is
<code><a href="../reference/extractMoreauBroto.html">extractMoreauBroto()</a></code>. A typical call would be:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">moreau</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractMoreauBroto.html">extractMoreauBroto</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">moreau</span>, n <span class="op">=</span> <span class="fl">36L</span><span class="op">)</span></span>
<span><span class="co">#&gt;  CIDH920105.lag1  CIDH920105.lag2  CIDH920105.lag3  CIDH920105.lag4 </span></span>
<span><span class="co">#&gt;      0.081573213     -0.016064817     -0.015982990     -0.025739038 </span></span>
<span><span class="co">#&gt;  CIDH920105.lag5  CIDH920105.lag6  CIDH920105.lag7  CIDH920105.lag8 </span></span>
<span><span class="co">#&gt;      0.079058632     -0.042771564     -0.036320847      0.024087298 </span></span>
<span><span class="co">#&gt;  CIDH920105.lag9 CIDH920105.lag10 CIDH920105.lag11 CIDH920105.lag12 </span></span>
<span><span class="co">#&gt;     -0.005273958      0.052274763      0.082170073      0.005419919 </span></span>
<span><span class="co">#&gt; CIDH920105.lag13 CIDH920105.lag14 CIDH920105.lag15 CIDH920105.lag16 </span></span>
<span><span class="co">#&gt;      0.083292042      0.004810584      0.001872446     -0.001531495 </span></span>
<span><span class="co">#&gt; CIDH920105.lag17 CIDH920105.lag18 CIDH920105.lag19 CIDH920105.lag20 </span></span>
<span><span class="co">#&gt;     -0.011917230      0.071161551      0.033473197      0.026882737 </span></span>
<span><span class="co">#&gt; CIDH920105.lag21 CIDH920105.lag22 CIDH920105.lag23 CIDH920105.lag24 </span></span>
<span><span class="co">#&gt;      0.073075402      0.115272790      0.041517897     -0.027025993 </span></span>
<span><span class="co">#&gt; CIDH920105.lag25 CIDH920105.lag26 CIDH920105.lag27 CIDH920105.lag28 </span></span>
<span><span class="co">#&gt;      0.033477388     -0.003245255      0.078117010     -0.028177304 </span></span>
<span><span class="co">#&gt; CIDH920105.lag29 CIDH920105.lag30  BHAR880101.lag1  BHAR880101.lag2 </span></span>
<span><span class="co">#&gt;      0.046695832      0.020584423      0.052740185      0.030804784 </span></span>
<span><span class="co">#&gt;  BHAR880101.lag3  BHAR880101.lag4  BHAR880101.lag5  BHAR880101.lag6 </span></span>
<span><span class="co">#&gt;      0.037170476     -0.058993771      0.070641780     -0.089192490</span></span></code></pre></div>
<p>The eight default properties used here are:</p>
<ul>
<li>AccNo. CIDH920105 - Normalized Average Hydrophobicity Scales</li>
<li>AccNo. BHAR880101 - Average Flexibility Indices</li>
<li>AccNo. CHAM820101 - Polarizability Parameter</li>
<li>AccNo. CHAM820102 - Free Energy of Solution in Water, kcal/mole</li>
<li>AccNo. CHOC760101 - Residue Accessible Surface Area in
Tripeptide</li>
<li>AccNo. BIGC670101 - Residue Volume</li>
<li>AccNo. CHAM810101 - Steric Parameter</li>
<li>AccNo. DAYM780201 - Relative Mutability</li>
</ul>
<p>Users can change the property names of AAindex database with the
argument <code>props</code>. The AAindex data shipped with protr can be
loaded by <code>data(AAindex)</code>, which has the detailed information
of each property. With the argument <code>customprops</code> and
<code>nlag</code>, users can specify their own properties and lag value
to calculate with. For example:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Define 3 custom properties</span></span>
<span><span class="va">myprops</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  AccNo <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MyProp1"</span>, <span class="st">"MyProp2"</span>, <span class="st">"MyProp3"</span><span class="op">)</span>,</span>
<span>  A <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.62</span>, <span class="op">-</span><span class="fl">0.5</span>, <span class="fl">15</span><span class="op">)</span>, R <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">2.53</span>, <span class="fl">3</span>, <span class="fl">101</span><span class="op">)</span>,</span>
<span>  N <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.78</span>, <span class="fl">0.2</span>, <span class="fl">58</span><span class="op">)</span>, D <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.9</span>, <span class="fl">3</span>, <span class="fl">59</span><span class="op">)</span>,</span>
<span>  C <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.29</span>, <span class="op">-</span><span class="fl">1</span>, <span class="fl">47</span><span class="op">)</span>, E <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.74</span>, <span class="fl">3</span>, <span class="fl">73</span><span class="op">)</span>,</span>
<span>  Q <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.85</span>, <span class="fl">0.2</span>, <span class="fl">72</span><span class="op">)</span>, G <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.48</span>, <span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>  H <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.4</span>, <span class="op">-</span><span class="fl">0.5</span>, <span class="fl">82</span><span class="op">)</span>, I <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.38</span>, <span class="op">-</span><span class="fl">1.8</span>, <span class="fl">57</span><span class="op">)</span>,</span>
<span>  L <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.06</span>, <span class="op">-</span><span class="fl">1.8</span>, <span class="fl">57</span><span class="op">)</span>, K <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">1.5</span>, <span class="fl">3</span>, <span class="fl">73</span><span class="op">)</span>,</span>
<span>  M <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.64</span>, <span class="op">-</span><span class="fl">1.3</span>, <span class="fl">75</span><span class="op">)</span>, F <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.19</span>, <span class="op">-</span><span class="fl">2.5</span>, <span class="fl">91</span><span class="op">)</span>,</span>
<span>  P <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.12</span>, <span class="fl">0</span>, <span class="fl">42</span><span class="op">)</span>, S <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.18</span>, <span class="fl">0.3</span>, <span class="fl">31</span><span class="op">)</span>,</span>
<span>  T <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">-</span><span class="fl">0.05</span>, <span class="op">-</span><span class="fl">0.4</span>, <span class="fl">45</span><span class="op">)</span>, W <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.81</span>, <span class="op">-</span><span class="fl">3.4</span>, <span class="fl">130</span><span class="op">)</span>,</span>
<span>  Y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.26</span>, <span class="op">-</span><span class="fl">2.3</span>, <span class="fl">107</span><span class="op">)</span>, V <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.08</span>, <span class="op">-</span><span class="fl">1.5</span>, <span class="fl">43</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use 4 properties in the AAindex database, and 3 cutomized properties</span></span>
<span><span class="va">moreau2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractMoreauBroto.html">extractMoreauBroto</a></span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  customprops <span class="op">=</span> <span class="va">myprops</span>,</span>
<span>  props <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"CIDH920105"</span>, <span class="st">"BHAR880101"</span>,</span>
<span>    <span class="st">"CHAM820101"</span>, <span class="st">"CHAM820102"</span>,</span>
<span>    <span class="st">"MyProp1"</span>, <span class="st">"MyProp2"</span>, <span class="st">"MyProp3"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">moreau2</span>, n <span class="op">=</span> <span class="fl">36L</span><span class="op">)</span></span>
<span><span class="co">#&gt;  CIDH920105.lag1  CIDH920105.lag2  CIDH920105.lag3  CIDH920105.lag4 </span></span>
<span><span class="co">#&gt;      0.081573213     -0.016064817     -0.015982990     -0.025739038 </span></span>
<span><span class="co">#&gt;  CIDH920105.lag5  CIDH920105.lag6  CIDH920105.lag7  CIDH920105.lag8 </span></span>
<span><span class="co">#&gt;      0.079058632     -0.042771564     -0.036320847      0.024087298 </span></span>
<span><span class="co">#&gt;  CIDH920105.lag9 CIDH920105.lag10 CIDH920105.lag11 CIDH920105.lag12 </span></span>
<span><span class="co">#&gt;     -0.005273958      0.052274763      0.082170073      0.005419919 </span></span>
<span><span class="co">#&gt; CIDH920105.lag13 CIDH920105.lag14 CIDH920105.lag15 CIDH920105.lag16 </span></span>
<span><span class="co">#&gt;      0.083292042      0.004810584      0.001872446     -0.001531495 </span></span>
<span><span class="co">#&gt; CIDH920105.lag17 CIDH920105.lag18 CIDH920105.lag19 CIDH920105.lag20 </span></span>
<span><span class="co">#&gt;     -0.011917230      0.071161551      0.033473197      0.026882737 </span></span>
<span><span class="co">#&gt; CIDH920105.lag21 CIDH920105.lag22 CIDH920105.lag23 CIDH920105.lag24 </span></span>
<span><span class="co">#&gt;      0.073075402      0.115272790      0.041517897     -0.027025993 </span></span>
<span><span class="co">#&gt; CIDH920105.lag25 CIDH920105.lag26 CIDH920105.lag27 CIDH920105.lag28 </span></span>
<span><span class="co">#&gt;      0.033477388     -0.003245255      0.078117010     -0.028177304 </span></span>
<span><span class="co">#&gt; CIDH920105.lag29 CIDH920105.lag30  BHAR880101.lag1  BHAR880101.lag2 </span></span>
<span><span class="co">#&gt;      0.046695832      0.020584423      0.052740185      0.030804784 </span></span>
<span><span class="co">#&gt;  BHAR880101.lag3  BHAR880101.lag4  BHAR880101.lag5  BHAR880101.lag6 </span></span>
<span><span class="co">#&gt;      0.037170476     -0.058993771      0.070641780     -0.089192490</span></span></code></pre></div>
<p>About the standard input format of <code>props</code> and
<code>customprops</code>, see <code><a href="../reference/extractMoreauBroto.html">?extractMoreauBroto</a></code> for
details.</p>
</div>
<div class="section level4">
<h4 id="moran-autocorrelation-descriptors">Moran autocorrelation descriptors<a class="anchor" aria-label="anchor" href="#moran-autocorrelation-descriptors"></a>
</h4>
<p>For protein sequences, the Moran autocorrelation descriptors can be
defined as:</p>
<p><span class="math display">\[
I(d) = \frac{\frac{1}{N-d} \sum_{i=1}^{N-d} (P_i - \bar{P}')
(P_{i+d} - \bar{P}')}{\frac{1}{N} \sum_{i=1}^{N} (P_i -
\bar{P}')^2} \quad d = 1, 2, \ldots, 30
\]</span></p>
<p>where <span class="math inline">\(d\)</span>, <span class="math inline">\(P_i\)</span>, and <span class="math inline">\(P_{i+d}\)</span> are defined in the same way as in
the first place; <span class="math inline">\(\bar{P}'\)</span> is
the considered property <span class="math inline">\(P\)</span> along the
sequence, i.e.,</p>
<p><span class="math display">\[
\bar{P}' = \frac{\sum_{i=1}^N P_i}{N}
\]</span></p>
<p><span class="math inline">\(d\)</span>, <span class="math inline">\(P\)</span>, <span class="math inline">\(P_i\)</span> and <span class="math inline">\(P_{i+d}\)</span>, <span class="math inline">\(\textrm{nlag}\)</span> have the same meaning as
above.</p>
<p>With <code><a href="../reference/extractMoran.html">extractMoran()</a></code> (which has the identical parameters
as <code><a href="../reference/extractMoreauBroto.html">extractMoreauBroto()</a></code>), we can compute the Moran
autocorrelation descriptors (only print out the first 36 elements):</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use the 3 custom properties defined before</span></span>
<span><span class="co"># and 4 properties in the AAindex database</span></span>
<span><span class="va">moran</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractMoran.html">extractMoran</a></span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  customprops <span class="op">=</span> <span class="va">myprops</span>,</span>
<span>  props <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"CIDH920105"</span>, <span class="st">"BHAR880101"</span>,</span>
<span>    <span class="st">"CHAM820101"</span>, <span class="st">"CHAM820102"</span>,</span>
<span>    <span class="st">"MyProp1"</span>, <span class="st">"MyProp2"</span>, <span class="st">"MyProp3"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">moran</span>, n <span class="op">=</span> <span class="fl">36L</span><span class="op">)</span></span>
<span><span class="co">#&gt;  CIDH920105.lag1  CIDH920105.lag2  CIDH920105.lag3  CIDH920105.lag4 </span></span>
<span><span class="co">#&gt;      0.062895724     -0.044827681     -0.045065117     -0.055955678 </span></span>
<span><span class="co">#&gt;  CIDH920105.lag5  CIDH920105.lag6  CIDH920105.lag7  CIDH920105.lag8 </span></span>
<span><span class="co">#&gt;      0.060586377     -0.074128412     -0.067308852     -0.001293384 </span></span>
<span><span class="co">#&gt;  CIDH920105.lag9 CIDH920105.lag10 CIDH920105.lag11 CIDH920105.lag12 </span></span>
<span><span class="co">#&gt;     -0.033747588      0.029392193      0.061789800     -0.023368437 </span></span>
<span><span class="co">#&gt; CIDH920105.lag13 CIDH920105.lag14 CIDH920105.lag15 CIDH920105.lag16 </span></span>
<span><span class="co">#&gt;      0.062769417     -0.024912264     -0.028298043     -0.031584063 </span></span>
<span><span class="co">#&gt; CIDH920105.lag17 CIDH920105.lag18 CIDH920105.lag19 CIDH920105.lag20 </span></span>
<span><span class="co">#&gt;     -0.043466730      0.047830694      0.005883901     -0.001769769 </span></span>
<span><span class="co">#&gt; CIDH920105.lag21 CIDH920105.lag22 CIDH920105.lag23 CIDH920105.lag24 </span></span>
<span><span class="co">#&gt;      0.049334048      0.096427969      0.015147594     -0.060092509 </span></span>
<span><span class="co">#&gt; CIDH920105.lag25 CIDH920105.lag26 CIDH920105.lag27 CIDH920105.lag28 </span></span>
<span><span class="co">#&gt;      0.007549152     -0.033987885      0.056307675     -0.061844453 </span></span>
<span><span class="co">#&gt; CIDH920105.lag29 CIDH920105.lag30  BHAR880101.lag1  BHAR880101.lag2 </span></span>
<span><span class="co">#&gt;      0.021484780     -0.008461776      0.014229951     -0.009142419 </span></span>
<span><span class="co">#&gt;  BHAR880101.lag3  BHAR880101.lag4  BHAR880101.lag5  BHAR880101.lag6 </span></span>
<span><span class="co">#&gt;     -0.003272262     -0.109613332      0.033346233     -0.141538598</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="geary-autocorrelation-descriptors">Geary autocorrelation descriptors<a class="anchor" aria-label="anchor" href="#geary-autocorrelation-descriptors"></a>
</h4>
<p>Geary autocorrelation descriptors for protein sequences can be
defined as:</p>
<p><span class="math display">\[
C(d) = \frac{\frac{1}{2(N-d)} \sum_{i=1}^{N-d} (P_i -
P_{i+d})^2}{\frac{1}{N-1} \sum_{i=1}^{N} (P_i - \bar{P}')^2} \quad d
= 1, 2, \ldots, 30
\]</span></p>
<p>where <span class="math inline">\(d\)</span>, <span class="math inline">\(P\)</span>, <span class="math inline">\(P_i\)</span>, <span class="math inline">\(P_{i+d}\)</span>, and <span class="math inline">\(\textrm{nlag}\)</span> have the same meaning as
above.</p>
<p>For each amino acid index, there will be <span class="math inline">\(3 \times \textrm{nlag}\)</span> autocorrelation
descriptors. The usage of <code><a href="../reference/extractGeary.html">extractGeary()</a></code> is exactly the
same as <code><a href="../reference/extractMoreauBroto.html">extractMoreauBroto()</a></code> and
<code><a href="../reference/extractMoran.html">extractMoran()</a></code>:</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Use the 3 custom properties defined before</span></span>
<span><span class="co"># and 4 properties in the AAindex database</span></span>
<span><span class="va">geary</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractGeary.html">extractGeary</a></span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  customprops <span class="op">=</span> <span class="va">myprops</span>,</span>
<span>  props <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"CIDH920105"</span>, <span class="st">"BHAR880101"</span>,</span>
<span>    <span class="st">"CHAM820101"</span>, <span class="st">"CHAM820102"</span>,</span>
<span>    <span class="st">"MyProp1"</span>, <span class="st">"MyProp2"</span>, <span class="st">"MyProp3"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">geary</span>, n <span class="op">=</span> <span class="fl">36L</span><span class="op">)</span></span>
<span><span class="co">#&gt;  CIDH920105.lag1  CIDH920105.lag2  CIDH920105.lag3  CIDH920105.lag4 </span></span>
<span><span class="co">#&gt;        0.9361830        1.0442920        1.0452843        1.0563467 </span></span>
<span><span class="co">#&gt;  CIDH920105.lag5  CIDH920105.lag6  CIDH920105.lag7  CIDH920105.lag8 </span></span>
<span><span class="co">#&gt;        0.9406031        1.0765517        1.0675786        0.9991363 </span></span>
<span><span class="co">#&gt;  CIDH920105.lag9 CIDH920105.lag10 CIDH920105.lag11 CIDH920105.lag12 </span></span>
<span><span class="co">#&gt;        1.0316555        0.9684585        0.9353130        1.0201990 </span></span>
<span><span class="co">#&gt; CIDH920105.lag13 CIDH920105.lag14 CIDH920105.lag15 CIDH920105.lag16 </span></span>
<span><span class="co">#&gt;        0.9340933        1.0207373        1.0251486        1.0290464 </span></span>
<span><span class="co">#&gt; CIDH920105.lag17 CIDH920105.lag18 CIDH920105.lag19 CIDH920105.lag20 </span></span>
<span><span class="co">#&gt;        1.0414375        0.9494403        0.9905987        0.9987183 </span></span>
<span><span class="co">#&gt; CIDH920105.lag21 CIDH920105.lag22 CIDH920105.lag23 CIDH920105.lag24 </span></span>
<span><span class="co">#&gt;        0.9472542        0.9010009        0.9828848        1.0574098 </span></span>
<span><span class="co">#&gt; CIDH920105.lag25 CIDH920105.lag26 CIDH920105.lag27 CIDH920105.lag28 </span></span>
<span><span class="co">#&gt;        0.9897955        1.0290018        0.9400066        1.0584150 </span></span>
<span><span class="co">#&gt; CIDH920105.lag29 CIDH920105.lag30  BHAR880101.lag1  BHAR880101.lag2 </span></span>
<span><span class="co">#&gt;        0.9762904        1.0029734        0.9818711        1.0051730 </span></span>
<span><span class="co">#&gt;  BHAR880101.lag3  BHAR880101.lag4  BHAR880101.lag5  BHAR880101.lag6 </span></span>
<span><span class="co">#&gt;        0.9967069        1.1012905        0.9595859        1.1337056</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="compositiontransitiondistribution">Composition/Transition/Distribution<a class="anchor" aria-label="anchor" href="#compositiontransitiondistribution"></a>
</h3>
<p>The CTD descriptors are developed by <span class="citation">Dubchak
<em>et al.</em> (1995)</span> and <span class="citation">Dubchak <em>et
al.</em> (1999)</span>.</p>
<img src="figures/CTD.png" width="100%" class="img-responsive" align="center" alt="CTD descriptors"><p align="left">
Figure 3: The sequence of a hypothetic protein indicating the
construction of composition, transition, and distribution descriptors of
a protein. Sequence index indicates the position of an amino acid in the
sequence. The index for each type of amino acids in the sequence
(<code>1</code>, <code>2</code> or <code>3</code>) indicates the
position of the first, second, third, â€¦ of that type of amino acid. 1/2
transition indicates the position of <code>12</code> or <code>21</code>
pairs in the sequence (1/3 and 2/3 are defined in the same way).
</p>
<p><strong>Step 1: Sequence Encoding</strong></p>
<p>The amino acids are categorized into three classes according to its
attribute, and each amino acid is encoded by one of the indices 1, 2, 3
according to which class it belongs. The attributes used here include
hydrophobicity, normalized van der Waals volume, polarity, and
polarizability. The corresponding classification for each attribute is
listed in Table 1.</p>
<table class="table">
<caption>Table 1: Amino acid attributes, and the three-group
classification of the 20 amino acids by each attribute</caption>
<thead><tr class="header">
<th></th>
<th align="left"><strong>Group 1</strong></th>
<th align="left"><strong>Group 2</strong></th>
<th align="left"><strong>Group 3</strong></th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Hydrophobicity</td>
<td align="left">Polar</td>
<td align="left">Neutral</td>
<td align="left">Hydrophobicity</td>
</tr>
<tr class="even">
<td></td>
<td align="left">R, K, E, D, Q, N</td>
<td align="left">G, A, S, T, P, H, Y</td>
<td align="left">C, L, V, I, M, F, W</td>
</tr>
<tr class="odd">
<td>Normalized van der Waals Volume</td>
<td align="left">0-2.78</td>
<td align="left">2.95-4.0</td>
<td align="left">4.03-8.08</td>
</tr>
<tr class="even">
<td></td>
<td align="left">G, A, S, T, P, D, C</td>
<td align="left">N, V, E, Q, I, L</td>
<td align="left">M, H, K, F, R, Y, W</td>
</tr>
<tr class="odd">
<td>Polarity</td>
<td align="left">4.9-6.2</td>
<td align="left">8.0-9.2</td>
<td align="left">10.4-13.0</td>
</tr>
<tr class="even">
<td></td>
<td align="left">L, I, F, W, C, M, V, Y</td>
<td align="left">P, A, T, G, S</td>
<td align="left">H, Q, R, K, N, E, D</td>
</tr>
<tr class="odd">
<td>Polarizability</td>
<td align="left">0-1.08</td>
<td align="left">0.128-0.186</td>
<td align="left">0.219-0.409</td>
</tr>
<tr class="even">
<td></td>
<td align="left">G, A, S, D, T</td>
<td align="left">C, P, N, V, E, Q, I, L</td>
<td align="left">K, M, H, F, R, Y, W</td>
</tr>
<tr class="odd">
<td>Charge</td>
<td align="left">Positive</td>
<td align="left">Neutral</td>
<td align="left">Negative</td>
</tr>
<tr class="even">
<td></td>
<td align="left">K, R</td>
<td align="left">A, N, C, Q, G, H, I, L, M, F, P, S, T, W, Y, V</td>
<td align="left">D, E</td>
</tr>
<tr class="odd">
<td>Secondary Structure</td>
<td align="left">Helix</td>
<td align="left">Strand</td>
<td align="left">Coil</td>
</tr>
<tr class="even">
<td></td>
<td align="left">E, A, L, M, Q, K, R, H</td>
<td align="left">V, I, Y, C, W, F, T</td>
<td align="left">G, N, P, S, D</td>
</tr>
<tr class="odd">
<td>Solvent Accessibility</td>
<td align="left">Buried</td>
<td align="left">Exposed</td>
<td align="left">Intermediate</td>
</tr>
<tr class="even">
<td></td>
<td align="left">A, L, F, C, G, I, V, W</td>
<td align="left">R, K, Q, E, N, D</td>
<td align="left">M, S, P, T, H, Y</td>
</tr>
</tbody>
</table>
<p>For example, for a given sequence <code>MTEITAAMVKELRESTGAGA</code>,
it will be encoded as <code>32132223311311222222</code> according to its
hydrophobicity.</p>
<p><strong>Step 2: Compute Composition, Transition, and Distribution
Descriptors</strong></p>
<p>Three types of descriptors, <em>Composition</em> (<em>C</em>),
<em>Transition</em> (<em>T</em>), and <em>Distribution</em> (<em>D</em>)
can be calculated for a given attribute as follows.</p>
<div class="section level4">
<h4 id="composition">Composition<a class="anchor" aria-label="anchor" href="#composition"></a>
</h4>
<p>Composition is defined as the global percentage for each encoded
class in the protein sequence. In the above example using the
hydrophobicity classification, the numbers for encoded classes
<code>1</code>, <code>2</code>, <code>3</code> are 5, 10, and 5, so that
the compositions for them will be <span class="math inline">\(5/20=25\%\)</span>, <span class="math inline">\(10/20=10\%\)</span>, and <span class="math inline">\(5/20=25\%\)</span>, where 20 is the length of the
protein sequence. The composition descriptor can be expressed as</p>
<p><span class="math display">\[
C_r = \frac{n_r}{n} \quad r = 1, 2, 3
\]</span></p>
<p>where <span class="math inline">\(n_r\)</span> is the number of amino
acid type <span class="math inline">\(r\)</span> in the encoded
sequence; <span class="math inline">\(N\)</span> is the length of the
sequence. An example for <code><a href="../reference/extractCTDC.html">extractCTDC()</a></code>:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extractCTDC.html">extractCTDC</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt;  hydrophobicity.Group1  hydrophobicity.Group2  hydrophobicity.Group3 </span></span>
<span><span class="co">#&gt;             0.29715302             0.40569395             0.29715302 </span></span>
<span><span class="co">#&gt; normwaalsvolume.Group1 normwaalsvolume.Group2 normwaalsvolume.Group3 </span></span>
<span><span class="co">#&gt;             0.45195730             0.29715302             0.25088968 </span></span>
<span><span class="co">#&gt;        polarity.Group1        polarity.Group2        polarity.Group3 </span></span>
<span><span class="co">#&gt;             0.33985765             0.33274021             0.32740214 </span></span>
<span><span class="co">#&gt;  polarizability.Group1  polarizability.Group2  polarizability.Group3 </span></span>
<span><span class="co">#&gt;             0.33096085             0.41814947             0.25088968 </span></span>
<span><span class="co">#&gt;          charge.Group1          charge.Group2          charge.Group3 </span></span>
<span><span class="co">#&gt;             0.11032028             0.79003559             0.09964413 </span></span>
<span><span class="co">#&gt; secondarystruct.Group1 secondarystruct.Group2 secondarystruct.Group3 </span></span>
<span><span class="co">#&gt;             0.38967972             0.29537367             0.31494662 </span></span>
<span><span class="co">#&gt;   solventaccess.Group1   solventaccess.Group2   solventaccess.Group3 </span></span>
<span><span class="co">#&gt;             0.43060498             0.29715302             0.27224199</span></span></code></pre></div>
<p>The result shows the elements who are named as
<code>PropertyNumber.GroupNumber</code> in the returned vector.</p>
</div>
<div class="section level4">
<h4 id="transition">Transition<a class="anchor" aria-label="anchor" href="#transition"></a>
</h4>
<p>A transition from class 1 to 2 is the percent frequency with which 1
is followed by 2 or 2 is followed by 1 in the encoded sequences. The
transition descriptor can be computed as</p>
<p><span class="math display">\[
T_{rs} = \frac{n_{rs} + n_{sr}}{N - 1} \quad rs = \text{12}, \text{13},
\text{23}
\]</span></p>
<p>where <span class="math inline">\(n_{rs}\)</span>, <span class="math inline">\(n_{sr}\)</span> are the numbers of dipeptide
encoded as <code>rs</code> and <code>sr</code> in the sequence; <span class="math inline">\(N\)</span> is the length of the sequence. An
example for <code><a href="../reference/extractCTDT.html">extractCTDT()</a></code>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extractCTDT.html">extractCTDT</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt; prop1.Tr1221 prop1.Tr1331 prop1.Tr2332 prop2.Tr1221 prop2.Tr1331 prop2.Tr2332 </span></span>
<span><span class="co">#&gt;   0.27094474   0.16042781   0.23351159   0.26737968   0.22638146   0.17112299 </span></span>
<span><span class="co">#&gt; prop3.Tr1221 prop3.Tr1331 prop3.Tr2332 prop4.Tr1221 prop4.Tr1331 prop4.Tr2332 </span></span>
<span><span class="co">#&gt;   0.21033868   0.20499109   0.23707665   0.27272727   0.15151515   0.24598930 </span></span>
<span><span class="co">#&gt; prop5.Tr1221 prop5.Tr1331 prop5.Tr2332 prop6.Tr1221 prop6.Tr1331 prop6.Tr2332 </span></span>
<span><span class="co">#&gt;   0.18181818   0.02139037   0.15686275   0.21925134   0.22816399   0.15864528 </span></span>
<span><span class="co">#&gt; prop7.Tr1221 prop7.Tr1331 prop7.Tr2332 </span></span>
<span><span class="co">#&gt;   0.25133690   0.21568627   0.18003565</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="distribution">Distribution<a class="anchor" aria-label="anchor" href="#distribution"></a>
</h4>
<p>The <em>distribution</em> descriptor describes the distribution of
each attribute in the sequence.</p>
<p>There are five â€œdistributionâ€ descriptors for each attribute and they
are the position percents in the whole sequence for the first residue,
25% residues, 50% residues, 75% residues, and 100% residues for a
certain encoded class. For example, there are 10 residues encoded as
<code>2</code> in the above example, the positions for the first residue
<code>2</code>, the 2nd residue <code>2</code> (25% * 10 = 2), the 5th
<code>2</code> residue (50% * 10 = 5), the 7th <code>2</code> (75% * 10
= 7) and the 10th residue <code>2</code> (100% * 10) in the encoded
sequence are 2, 5, 15, 17, 20, so that the distribution descriptors for
<code>2</code> are: 10.0 (2/20 * 100), 25.0 (5/20 * 100), 75.0 (15/20 *
100), 85.0 (17/20 * 100), 100.0 (20/20 * 100).</p>
<p>To compute the distribution descriptor, use
<code><a href="../reference/extractCTDD.html">extractCTDD()</a></code>:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extractCTDD.html">extractCTDD</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt;   prop1.G1.residue0  prop1.G1.residue25  prop1.G1.residue50  prop1.G1.residue75 </span></span>
<span><span class="co">#&gt;           0.3558719          23.1316726          50.1779359          73.8434164 </span></span>
<span><span class="co">#&gt; prop1.G1.residue100   prop1.G2.residue0  prop1.G2.residue25  prop1.G2.residue50 </span></span>
<span><span class="co">#&gt;          99.8220641           0.5338078          27.4021352          47.3309609 </span></span>
<span><span class="co">#&gt;  prop1.G2.residue75 prop1.G2.residue100   prop1.G3.residue0  prop1.G3.residue25 </span></span>
<span><span class="co">#&gt;          75.2669039         100.0000000           0.1779359          19.5729537 </span></span>
<span><span class="co">#&gt;  prop1.G3.residue50  prop1.G3.residue75 prop1.G3.residue100   prop2.G1.residue0 </span></span>
<span><span class="co">#&gt;          51.7793594          75.6227758          99.6441281           0.3558719 </span></span>
<span><span class="co">#&gt;  prop2.G1.residue25  prop2.G1.residue50  prop2.G1.residue75 prop2.G1.residue100 </span></span>
<span><span class="co">#&gt;          25.6227758          48.0427046          75.4448399         100.0000000 </span></span>
<span><span class="co">#&gt;   prop2.G2.residue0  prop2.G2.residue25  prop2.G2.residue50  prop2.G2.residue75 </span></span>
<span><span class="co">#&gt;           1.4234875          23.3096085          54.4483986          76.3345196 </span></span>
<span><span class="co">#&gt; prop2.G2.residue100   prop2.G3.residue0  prop2.G3.residue25  prop2.G3.residue50 </span></span>
<span><span class="co">#&gt;          99.4661922           0.1779359          22.7758007          48.9323843 </span></span>
<span><span class="co">#&gt;  prop2.G3.residue75 prop2.G3.residue100   prop3.G1.residue0  prop3.G1.residue25 </span></span>
<span><span class="co">#&gt;          69.5729537          99.8220641           0.1779359          20.9964413 </span></span>
<span><span class="co">#&gt;  prop3.G1.residue50  prop3.G1.residue75 prop3.G1.residue100   prop3.G2.residue0 </span></span>
<span><span class="co">#&gt;          50.8896797          74.5551601          99.6441281           0.5338078 </span></span>
<span><span class="co">#&gt;  prop3.G2.residue25  prop3.G2.residue50  prop3.G2.residue75 prop3.G2.residue100 </span></span>
<span><span class="co">#&gt;          26.5124555          46.2633452          75.4448399         100.0000000 </span></span>
<span><span class="co">#&gt;   prop3.G3.residue0  prop3.G3.residue25  prop3.G3.residue50  prop3.G3.residue75 </span></span>
<span><span class="co">#&gt;           0.3558719          24.1992883          50.5338078          73.8434164 </span></span>
<span><span class="co">#&gt; prop3.G3.residue100   prop4.G1.residue0  prop4.G1.residue25  prop4.G1.residue50 </span></span>
<span><span class="co">#&gt;          99.8220641           0.3558719          26.5124555          48.3985765 </span></span>
<span><span class="co">#&gt;  prop4.G1.residue75 prop4.G1.residue100   prop4.G2.residue0  prop4.G2.residue25 </span></span>
<span><span class="co">#&gt;          76.1565836          99.2882562           1.4234875          21.5302491 </span></span>
<span><span class="co">#&gt;  prop4.G2.residue50  prop4.G2.residue75 prop4.G2.residue100   prop4.G3.residue0 </span></span>
<span><span class="co">#&gt;          51.4234875          75.8007117         100.0000000           0.1779359 </span></span>
<span><span class="co">#&gt;  prop4.G3.residue25  prop4.G3.residue50  prop4.G3.residue75 prop4.G3.residue100 </span></span>
<span><span class="co">#&gt;          22.7758007          48.9323843          69.5729537          99.8220641 </span></span>
<span><span class="co">#&gt;   prop5.G1.residue0  prop5.G1.residue25  prop5.G1.residue50  prop5.G1.residue75 </span></span>
<span><span class="co">#&gt;           0.8896797          20.8185053          48.9323843          69.5729537 </span></span>
<span><span class="co">#&gt; prop5.G1.residue100   prop5.G2.residue0  prop5.G2.residue25  prop5.G2.residue50 </span></span>
<span><span class="co">#&gt;          99.8220641           0.1779359          24.9110320          49.1103203 </span></span>
<span><span class="co">#&gt;  prop5.G2.residue75 prop5.G2.residue100   prop5.G3.residue0  prop5.G3.residue25 </span></span>
<span><span class="co">#&gt;          75.2669039         100.0000000           0.3558719          26.1565836 </span></span>
<span><span class="co">#&gt;  prop5.G3.residue50  prop5.G3.residue75 prop5.G3.residue100   prop6.G1.residue0 </span></span>
<span><span class="co">#&gt;          64.2348754          77.4021352          99.2882562           0.1779359 </span></span>
<span><span class="co">#&gt;  prop6.G1.residue25  prop6.G1.residue50  prop6.G1.residue75 prop6.G1.residue100 </span></span>
<span><span class="co">#&gt;          22.9537367          50.8896797          74.3772242          99.8220641 </span></span>
<span><span class="co">#&gt;   prop6.G2.residue0  prop6.G2.residue25  prop6.G2.residue50  prop6.G2.residue75 </span></span>
<span><span class="co">#&gt;           1.6014235          21.5302491          49.2882562          70.8185053 </span></span>
<span><span class="co">#&gt; prop6.G2.residue100   prop6.G3.residue0  prop6.G3.residue25  prop6.G3.residue50 </span></span>
<span><span class="co">#&gt;          98.9323843           0.3558719          29.0035587          48.2206406 </span></span>
<span><span class="co">#&gt;  prop6.G3.residue75 prop6.G3.residue100   prop7.G1.residue0  prop7.G1.residue25 </span></span>
<span><span class="co">#&gt;          77.4021352         100.0000000           0.5338078          23.4875445 </span></span>
<span><span class="co">#&gt;  prop7.G1.residue50  prop7.G1.residue75 prop7.G1.residue100   prop7.G2.residue0 </span></span>
<span><span class="co">#&gt;          50.0000000          74.5551601          98.9323843           0.3558719 </span></span>
<span><span class="co">#&gt;  prop7.G2.residue25  prop7.G2.residue50  prop7.G2.residue75 prop7.G2.residue100 </span></span>
<span><span class="co">#&gt;          23.1316726          50.1779359          73.8434164          99.8220641 </span></span>
<span><span class="co">#&gt;   prop7.G3.residue0  prop7.G3.residue25  prop7.G3.residue50  prop7.G3.residue75 </span></span>
<span><span class="co">#&gt;           0.1779359          27.2241993          48.0427046          75.4448399 </span></span>
<span><span class="co">#&gt; prop7.G3.residue100 </span></span>
<span><span class="co">#&gt;         100.0000000</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="conjoint-triad-descriptors">Conjoint triad descriptors<a class="anchor" aria-label="anchor" href="#conjoint-triad-descriptors"></a>
</h3>
<p>Conjoint triad descriptors are proposed by <span class="citation">Shen <em>et al.</em> (2007)</span>. The conjoint triad
descriptors were used to model protein-protein interactions based on the
classification of amino acids. In this approach, each protein sequence
is represented by a vector space consisting of descriptors of amino
acids. To reduce the dimensions of vector space, the 20 amino acids were
clustered into several classes according to their dipoles and volumes of
the side chains. The conjoint triad descriptors are calculated as
follows:</p>
<p><strong>Step 1: Classification of Amino Acids</strong></p>
<p>Electrostatic and hydrophobic interactions dominate protein-protein
interactions. These two kinds of interactions may be reflected by the
dipoles and volumes of the side chains of amino acids, respectively.
Accordingly, these two parameters were calculated by using the
density-functional theory method B3LYP/6-31G and molecular modeling
approach. Based on the dipoles and volumes of the side chains, the 20
amino acids can be clustered into seven classes (See Table 2). Amino
acids within the same class likely involve synonymous mutations because
of their similar characteristics.</p>
<table class="table">
<caption>Table 2: Classification of amino acids based on dipoles and
volumes of the side chains</caption>
<thead><tr class="header">
<th align="center">No.</th>
<th align="center">Dipole Scale<span class="math inline">\(^1\)</span>
</th>
<th align="center">Volume Scale<span class="math inline">\(^2\)</span>
</th>
<th align="center">Class</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">1</td>
<td align="center"><span class="math inline">\(-\)</span></td>
<td align="center"><span class="math inline">\(-\)</span></td>
<td align="center">Ala, Gly, Val</td>
</tr>
<tr class="even">
<td align="center">2</td>
<td align="center"><span class="math inline">\(-\)</span></td>
<td align="center"><span class="math inline">\(+\)</span></td>
<td align="center">Ile, Leu, Phe, Pro</td>
</tr>
<tr class="odd">
<td align="center">3</td>
<td align="center"><span class="math inline">\(+\)</span></td>
<td align="center"><span class="math inline">\(+\)</span></td>
<td align="center">Tyr, Met, Thr, Ser</td>
</tr>
<tr class="even">
<td align="center">4</td>
<td align="center"><span class="math inline">\(++\)</span></td>
<td align="center"><span class="math inline">\(+\)</span></td>
<td align="center">His, Asn, Gln, Tpr</td>
</tr>
<tr class="odd">
<td align="center">5</td>
<td align="center"><span class="math inline">\(+++\)</span></td>
<td align="center"><span class="math inline">\(+\)</span></td>
<td align="center">Arg, Lys</td>
</tr>
<tr class="even">
<td align="center">6</td>
<td align="center"><span class="math inline">\(+'+'+'\)</span></td>
<td align="center"><span class="math inline">\(+\)</span></td>
<td align="center">Asp, Glu</td>
</tr>
<tr class="odd">
<td align="center">7</td>
<td align="center"><span class="math inline">\(+^3\)</span></td>
<td align="center"><span class="math inline">\(+\)</span></td>
<td align="center">Cys</td>
</tr>
</tbody>
</table>
<p align="left">
<sup>1</sup> Dipole Scale (Debye): <span class="math inline">\(-\)</span>, Dipole &lt; 1.0; <span class="math inline">\(+\)</span>, 1.0 &lt; Dipole &lt; 2.0; <span class="math inline">\(++\)</span>, 2.0 &lt; Dipole &lt; 3.0; <span class="math inline">\(+++\)</span>, Dipole &gt; 3.0; <span class="math inline">\(+'+'+'\)</span>, Dipole &gt; 3.0 with
opposite orientation.
</p>
<p align="left">
<sup>2</sup> Volume Scale (<span class="math inline">\(\overset{\lower.5em\circ}{\mathrm{A}}\lower.01em^3\)</span>):
<span class="math inline">\(-\)</span>, Volume &lt; 50; <span class="math inline">\(+\)</span>, Volume &gt; 50.
</p>
<p align="left">
<sup>3</sup> Cys is separated from class 3 because of its ability to
form disulfide bonds.
</p>
<p><strong>Step 2: Conjoint Triad Calculation</strong></p>
<p>The conjoint triad descriptors considered the properties of one amino
acid and its vicinal amino acids and regarded any three continuous amino
acids as a unit. Thus, the triads can be differentiated according to the
classes of amino acids, i.e., triads composed by three amino acids
belonging to the same classes, such as ART and VKS, can be treated
identically. To conveniently represent a protein, we first use a binary
space <span class="math inline">\((\mathbf{V}, \mathbf{F})\)</span> to
represent a protein sequence. Here, <span class="math inline">\(\mathbf{V}\)</span> is the vector space of the
sequence features, and each feature <span class="math inline">\(v_i\)</span> represents a sort of triad type;
<span class="math inline">\(\mathbf{F}\)</span> is the frequency vector
corresponding to <span class="math inline">\(\mathbf{V}\)</span>, and
the value of the <span class="math inline">\(i\)</span>-th dimension of
<span class="math inline">\(\mathbf{F} (f_i)\)</span> is the frequency
of type <span class="math inline">\(v_i\)</span> appearing in the
protein sequence. For the amino acids that have been catogorized into
seven classes, the size of <span class="math inline">\(\mathbf{V}\)</span> should be <span class="math inline">\(7 \times 7 \times 7\)</span>; thus <span class="math inline">\(i = 1, 2, \ldots, 343\)</span>. The detailed
description for (<span class="math inline">\(\mathbf{V}\)</span>, <span class="math inline">\(\mathbf{F}\)</span>) is illustrated in Figure
4.</p>
<img src="figures/ctriad.png" width="100%" class="img-responsive" align="center"><p align="left">
Figure 4: Schematic diagram for constructing the vector space (<span class="math inline">\(\mathbf{V}\)</span>, <span class="math inline">\(\mathbf{F}\)</span>) of protein sequences. <span class="math inline">\(\mathbf{V}\)</span> is the vector space of the
sequence features; each feature (<span class="math inline">\(v_i\)</span>) represents a triad composed of three
consecutive amino acids; <span class="math inline">\(\mathbf{F}\)</span>
is the frequency vector corresponding to <span class="math inline">\(\mathbf{V}\)</span>, and the value of the <span class="math inline">\(i\)</span>-th dimension of <span class="math inline">\(\mathbf{F} (f_i)\)</span> is the frequency that
<span class="math inline">\(v_i\)</span> triad appeared in the protein
sequence.
</p>
<p>Clearly, each protein correlates to the length (number of amino
acids) of protein. In general, a long protein would have a large value
of <span class="math inline">\(f_i\)</span>, which complicates the
comparison between two heterogeneous proteins. Thus, we defined a new
parameter, <span class="math inline">\(d_i\)</span>, by normalizing
<span class="math inline">\(f_i\)</span> with the following
equation:</p>
<p><span class="math display">\[
d_i = \frac{f_i - \min\{\,f_1, f_2 , \ldots, f_{343}\,\}}{\max\{\,f_1,
f_2, \ldots, f_{343}\,\}}
\]</span></p>
<p>The numerical value of <span class="math inline">\(d_i\)</span> of
each protein ranges from 0 to 1, which thereby enables the comparison
between proteins. Accordingly, we obtain another vector space
(designated <span class="math inline">\(\mathbf{D}\)</span>) consisting
of <span class="math inline">\(d_i\)</span> to represent protein.</p>
<p>To compute conjoint triads of protein sequences, we can simply
use:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ctriad</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractCTriad.html">extractCTriad</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">ctriad</span>, n <span class="op">=</span> <span class="fl">65L</span><span class="op">)</span></span>
<span><span class="co">#&gt; VS111 VS211 VS311 VS411 VS511 VS611 VS711 VS121 VS221 VS321 VS421 VS521 VS621 </span></span>
<span><span class="co">#&gt;   0.1   0.3   0.6   0.2   0.4   0.0   0.3   1.0   0.6   0.5   0.0   0.2   0.3 </span></span>
<span><span class="co">#&gt; VS721 VS131 VS231 VS331 VS431 VS531 VS631 VS731 VS141 VS241 VS341 VS441 VS541 </span></span>
<span><span class="co">#&gt;   0.0   0.2   0.4   0.5   0.2   0.3   0.3   0.1   0.3   0.3   0.2   0.2   0.0 </span></span>
<span><span class="co">#&gt; VS641 VS741 VS151 VS251 VS351 VS451 VS551 VS651 VS751 VS161 VS261 VS361 VS461 </span></span>
<span><span class="co">#&gt;   0.1   0.2   0.2   0.2   0.5   0.1   0.2   0.0   0.0   0.1   0.4   0.2   0.3 </span></span>
<span><span class="co">#&gt; VS561 VS661 VS761 VS171 VS271 VS371 VS471 VS571 VS671 VS771 VS112 VS212 VS312 </span></span>
<span><span class="co">#&gt;   0.2   0.0   0.1   0.1   0.3   0.1   0.0   0.1   0.0   0.1   0.8   0.4   0.4 </span></span>
<span><span class="co">#&gt; VS412 VS512 VS612 VS712 VS122 VS222 VS322 VS422 VS522 VS622 VS722 VS132 VS232 </span></span>
<span><span class="co">#&gt;   0.6   0.1   0.5   0.2   0.8   0.5   0.2   0.3   0.2   0.0   0.2   0.1   0.3</span></span></code></pre></div>
<p>by which we only outputted the first 65 of total 343 dimension to
save space.</p>
</div>
<div class="section level3">
<h3 id="quasi-sequence-order-descriptors">Quasi-sequence-order descriptors<a class="anchor" aria-label="anchor" href="#quasi-sequence-order-descriptors"></a>
</h3>
<p>The quasi-sequence-order descriptors are proposed by <span class="citation">Chou (2000)</span>. They are derived from the distance
matrix between the 20 amino acids.</p>
<div class="section level4">
<h4 id="sequence-order-coupling-number">Sequence-order-coupling number<a class="anchor" aria-label="anchor" href="#sequence-order-coupling-number"></a>
</h4>
<p>The <span class="math inline">\(d\)</span>-th rank
sequence-order-coupling number is defined as:</p>
<p><span class="math display">\[
\tau_d = \sum_{i=1}^{N-d} (d_{i, i+d})^2 \quad d = 1, 2, \ldots,
\textrm{maxlag}
\]</span></p>
<p>where <span class="math inline">\(d_{i, i+d}\)</span> is the distance
between the two amino acids at position <span class="math inline">\(i\)</span> and <span class="math inline">\(i+d\)</span>.</p>
<p><strong>Note</strong>: maxlag is the maximum lag and the length of
the protein must be not less than <span class="math inline">\(\textrm{maxlag}\)</span>.</p>
<p>The function <code><a href="../reference/extractSOCN.html">extractSOCN()</a></code> is used for computing the
sequence-order-coupling numbers:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extractSOCN.html">extractSOCN</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt;  Schneider.lag1  Schneider.lag2  Schneider.lag3  Schneider.lag4  Schneider.lag5 </span></span>
<span><span class="co">#&gt;        204.2036        199.8708        206.8102        197.4828        193.3366 </span></span>
<span><span class="co">#&gt;  Schneider.lag6  Schneider.lag7  Schneider.lag8  Schneider.lag9 Schneider.lag10 </span></span>
<span><span class="co">#&gt;        208.1936        195.5476        200.9789        196.7110        193.9931 </span></span>
<span><span class="co">#&gt; Schneider.lag11 Schneider.lag12 Schneider.lag13 Schneider.lag14 Schneider.lag15 </span></span>
<span><span class="co">#&gt;        199.7031        204.9389        187.0140        198.4702        205.4526 </span></span>
<span><span class="co">#&gt; Schneider.lag16 Schneider.lag17 Schneider.lag18 Schneider.lag19 Schneider.lag20 </span></span>
<span><span class="co">#&gt;        193.1274        187.3529        190.4949        202.8853        198.5299 </span></span>
<span><span class="co">#&gt; Schneider.lag21 Schneider.lag22 Schneider.lag23 Schneider.lag24 Schneider.lag25 </span></span>
<span><span class="co">#&gt;        191.1013        185.0074        189.9857        202.7113        201.6267 </span></span>
<span><span class="co">#&gt; Schneider.lag26 Schneider.lag27 Schneider.lag28 Schneider.lag29 Schneider.lag30 </span></span>
<span><span class="co">#&gt;        194.5770        185.9939        204.1297        191.1629        183.9073 </span></span>
<span><span class="co">#&gt;   Grantham.lag1   Grantham.lag2   Grantham.lag3   Grantham.lag4   Grantham.lag5 </span></span>
<span><span class="co">#&gt;    6674686.0000    6761609.0000    7138892.0000    6748261.0000    6291229.0000 </span></span>
<span><span class="co">#&gt;   Grantham.lag6   Grantham.lag7   Grantham.lag8   Grantham.lag9  Grantham.lag10 </span></span>
<span><span class="co">#&gt;    6839853.0000    6594164.0000    6556148.0000    6620183.0000    6770614.0000 </span></span>
<span><span class="co">#&gt;  Grantham.lag11  Grantham.lag12  Grantham.lag13  Grantham.lag14  Grantham.lag15 </span></span>
<span><span class="co">#&gt;    6495689.0000    6865537.0000    6297267.0000    6498247.0000    6615566.0000 </span></span>
<span><span class="co">#&gt;  Grantham.lag16  Grantham.lag17  Grantham.lag18  Grantham.lag19  Grantham.lag20 </span></span>
<span><span class="co">#&gt;    6572680.0000    6569081.0000    6173947.0000    6570829.0000    6471308.0000 </span></span>
<span><span class="co">#&gt;  Grantham.lag21  Grantham.lag22  Grantham.lag23  Grantham.lag24  Grantham.lag25 </span></span>
<span><span class="co">#&gt;    6461649.0000    5939432.0000    6532121.0000    6652472.0000    6480660.0000 </span></span>
<span><span class="co">#&gt;  Grantham.lag26  Grantham.lag27  Grantham.lag28  Grantham.lag29  Grantham.lag30 </span></span>
<span><span class="co">#&gt;    6382281.0000    6276521.0000    6537634.0000    6442991.0000    6350157.0000</span></span></code></pre></div>
<p>Users can also specify the maximum lag value with the
<code>nlag</code> argument.</p>
<p><strong>Note</strong>: In addition to Schneider-Wrede physicochemical
distance matrix <span class="citation">(Schneider and Wrede,
1994)</span> used by Kuo-Chen Chou, another chemical distance matrix by
<span class="citation">Grantham (1974)</span> is also used here. So the
descriptors dimension will be <code>nlag * 2</code>. The
quasi-sequence-order descriptors described next also utilized the two
matrices.</p>
</div>
<div class="section level4">
<h4 id="quasi-sequence-order-descriptors-1">Quasi-sequence-order descriptors<a class="anchor" aria-label="anchor" href="#quasi-sequence-order-descriptors-1"></a>
</h4>
<p>For each amino acid type, a quasi-sequence-order descriptor can be
defined as:</p>
<p><span class="math display">\[
X_r = \frac{f_r}{\sum_{r=1}^{20} f_r + w \sum_{d=1}^{\textrm{maxlag}}
\tau_d} \quad r = 1, 2, \ldots, 20
\]</span></p>
<p>where <span class="math inline">\(f_r\)</span> is the normalized
occurrence for amino acid type <span class="math inline">\(i\)</span>
and <span class="math inline">\(w\)</span> is a weighting factor (<span class="math inline">\(w=0.1\)</span>). These are the first 20
quasi-sequence-order descriptors. The other 30 quasi-sequence-order are
defined as:</p>
<p><span class="math display">\[
X_d = \frac{w \tau_{d-20}}{\sum_{r=1}^{20} f_r + w
\sum_{d=1}^{\textrm{maxlag}} \tau_d} \quad d = 21, 22, \ldots, 20 +
\textrm{maxlag}
\]</span></p>
<img src="figures/QSO.png" width="75%" class="img-responsive" align="center"><p align="left">
Figure 5: A schematic drawing to show (a) the 1st-rank, (b) the
2nd-rank, and (3) the 3rd-rank sequence-order-coupling mode along a
protein sequence. (a) Reflects the coupling mode between all the most
contiguous residues, (b) that between all the 2nd most contiguous
residues, and (c) that between all the 3rd most contiguous residues.
This figure is from <span class="citation">Chou (2000)</span>.
</p>
<p>A minimal example for <code><a href="../reference/extractQSO.html">extractQSO()</a></code>:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extractQSO.html">extractQSO</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt;  Schneider.Xr.A  Schneider.Xr.R  Schneider.Xr.N  Schneider.Xr.D  Schneider.Xr.C </span></span>
<span><span class="co">#&gt;    6.096218e-02    6.773576e-02    3.725467e-02    4.910842e-02    6.434897e-02 </span></span>
<span><span class="co">#&gt;  Schneider.Xr.E  Schneider.Xr.Q  Schneider.Xr.G  Schneider.Xr.H  Schneider.Xr.I </span></span>
<span><span class="co">#&gt;    4.572164e-02    4.572164e-02    7.789612e-02    2.878770e-02    3.386788e-02 </span></span>
<span><span class="co">#&gt;  Schneider.Xr.L  Schneider.Xr.K  Schneider.Xr.M  Schneider.Xr.F  Schneider.Xr.P </span></span>
<span><span class="co">#&gt;    7.281594e-02    3.725467e-02    1.185376e-02    3.048109e-02    5.080182e-02 </span></span>
<span><span class="co">#&gt;  Schneider.Xr.S  Schneider.Xr.T  Schneider.Xr.W  Schneider.Xr.Y  Schneider.Xr.V </span></span>
<span><span class="co">#&gt;    8.466970e-02    4.233485e-02    2.201412e-02    4.064145e-02    4.741503e-02 </span></span>
<span><span class="co">#&gt;   Grantham.Xr.A   Grantham.Xr.R   Grantham.Xr.N   Grantham.Xr.D   Grantham.Xr.C </span></span>
<span><span class="co">#&gt;    1.835033e-06    2.038926e-06    1.121409e-06    1.478221e-06    1.936980e-06 </span></span>
<span><span class="co">#&gt;   Grantham.Xr.E   Grantham.Xr.Q   Grantham.Xr.G   Grantham.Xr.H   Grantham.Xr.I </span></span>
<span><span class="co">#&gt;    1.376275e-06    1.376275e-06    2.344765e-06    8.665435e-07    1.019463e-06 </span></span>
<span><span class="co">#&gt;   Grantham.Xr.L   Grantham.Xr.K   Grantham.Xr.M   Grantham.Xr.F   Grantham.Xr.P </span></span>
<span><span class="co">#&gt;    2.191845e-06    1.121409e-06    3.568120e-07    9.175167e-07    1.529194e-06 </span></span>
<span><span class="co">#&gt;   Grantham.Xr.S   Grantham.Xr.T   Grantham.Xr.W   Grantham.Xr.Y   Grantham.Xr.V </span></span>
<span><span class="co">#&gt;    2.548657e-06    1.274329e-06    6.626509e-07    1.223356e-06    1.427248e-06 </span></span>
<span><span class="co">#&gt;  Schneider.Xd.1  Schneider.Xd.2  Schneider.Xd.3  Schneider.Xd.4  Schneider.Xd.5 </span></span>
<span><span class="co">#&gt;    3.457972e-02    3.384600e-02    3.502111e-02    3.344162e-02    3.273951e-02 </span></span>
<span><span class="co">#&gt;  Schneider.Xd.6  Schneider.Xd.7  Schneider.Xd.8  Schneider.Xd.9 Schneider.Xd.10 </span></span>
<span><span class="co">#&gt;    3.525537e-02    3.311390e-02    3.403364e-02    3.331093e-02    3.285068e-02 </span></span>
<span><span class="co">#&gt; Schneider.Xd.11 Schneider.Xd.12 Schneider.Xd.13 Schneider.Xd.14 Schneider.Xd.15 </span></span>
<span><span class="co">#&gt;    3.381760e-02    3.470422e-02    3.166883e-02    3.360882e-02    3.479121e-02 </span></span>
<span><span class="co">#&gt; Schneider.Xd.16 Schneider.Xd.17 Schneider.Xd.18 Schneider.Xd.19 Schneider.Xd.20 </span></span>
<span><span class="co">#&gt;    3.270408e-02    3.172623e-02    3.225829e-02    3.435647e-02    3.361893e-02 </span></span>
<span><span class="co">#&gt; Schneider.Xd.21 Schneider.Xd.22 Schneider.Xd.23 Schneider.Xd.24 Schneider.Xd.25 </span></span>
<span><span class="co">#&gt;    3.236099e-02    3.132904e-02    3.217206e-02    3.432701e-02    3.414334e-02 </span></span>
<span><span class="co">#&gt; Schneider.Xd.26 Schneider.Xd.27 Schneider.Xd.28 Schneider.Xd.29 Schneider.Xd.30 </span></span>
<span><span class="co">#&gt;    3.294954e-02    3.149609e-02    3.456720e-02    3.237140e-02    3.114275e-02 </span></span>
<span><span class="co">#&gt;   Grantham.Xd.1   Grantham.Xd.2   Grantham.Xd.3   Grantham.Xd.4   Grantham.Xd.5 </span></span>
<span><span class="co">#&gt;    3.402298e-02    3.446605e-02    3.638918e-02    3.439801e-02    3.206838e-02 </span></span>
<span><span class="co">#&gt;   Grantham.Xd.6   Grantham.Xd.7   Grantham.Xd.8   Grantham.Xd.9  Grantham.Xd.10 </span></span>
<span><span class="co">#&gt;    3.486488e-02    3.361253e-02    3.341875e-02    3.374516e-02    3.451195e-02 </span></span>
<span><span class="co">#&gt;  Grantham.Xd.11  Grantham.Xd.12  Grantham.Xd.13  Grantham.Xd.14  Grantham.Xd.15 </span></span>
<span><span class="co">#&gt;    3.311057e-02    3.499580e-02    3.209915e-02    3.312361e-02    3.372162e-02 </span></span>
<span><span class="co">#&gt;  Grantham.Xd.16  Grantham.Xd.17  Grantham.Xd.18  Grantham.Xd.19  Grantham.Xd.20 </span></span>
<span><span class="co">#&gt;    3.350302e-02    3.348467e-02    3.147055e-02    3.349358e-02    3.298629e-02 </span></span>
<span><span class="co">#&gt;  Grantham.Xd.21  Grantham.Xd.22  Grantham.Xd.23  Grantham.Xd.24  Grantham.Xd.25 </span></span>
<span><span class="co">#&gt;    3.293706e-02    3.027516e-02    3.329628e-02    3.390974e-02    3.303396e-02 </span></span>
<span><span class="co">#&gt;  Grantham.Xd.26  Grantham.Xd.27  Grantham.Xd.28  Grantham.Xd.29  Grantham.Xd.30 </span></span>
<span><span class="co">#&gt;    3.253250e-02    3.199340e-02    3.332438e-02    3.284195e-02    3.236875e-02</span></span></code></pre></div>
<p>where users can also specify the maximum lag with the argument
<code>nlag</code> and the weighting factor with the argument
<code>w</code>.</p>
</div>
</div>
<div class="section level3">
<h3 id="pseudo-amino-acid-composition-pseaac">Pseudo-amino acid composition (PseAAC)<a class="anchor" aria-label="anchor" href="#pseudo-amino-acid-composition-pseaac"></a>
</h3>
<p>This group of descriptors are proposed by <span class="citation">Chou
(2001)</span>. PseAAC descriptors are also named as the <em>type 1
pseudo-amino acid composition</em>. Let <span class="math inline">\(H_1^o (i)\)</span>, <span class="math inline">\(H_2^o (i)\)</span>, <span class="math inline">\(M^o (i)\)</span> (<span class="math inline">\(i=1,
2, 3, \ldots, 20\)</span>) be the original hydrophobicity values, the
original hydrophilicity values and the original side chain masses of the
20 natural amino acids, respectively. They are converted to following
qualities by a standard conversion:</p>
<p><span class="math display">\[
H_1 (i) = \frac{H_1^o (i) - \frac{1}{20} \sum_{i=1}^{20} H_1^o
(i)}{\sqrt{\frac{\sum_{i=1}^{20} [H_1^o (i) - \frac{1}{20}
\sum_{i=1}^{20} H_1^o (i) ]^2}{20}}}
\]</span></p>
<p><span class="math inline">\(H_2^o (i)\)</span> and <span class="math inline">\(M^o (i)\)</span> are normalized as <span class="math inline">\(H_2 (i)\)</span> and <span class="math inline">\(M
(i)\)</span> in the same way.</p>
<img src="figures/PAAC.png" width="75%" class="img-responsive" align="center"><p align="left">
Figure 6: A schematic drawing to show (a) the first-tier, (b) the
second-tier, and (3) the third-tiersequence order correlation mode along
a protein sequence. Panel (a) reflects the correlation mode between all
the most contiguous residues, panel (b) that between all the second-most
contiguous residues, and panel (c) that between all the third-most
contiguous residues. This figure is from <span class="citation">Chou
(2001)</span>.
</p>
<p>Then, a correlation function can be defines as</p>
<p><span class="math display">\[
\Theta (R_i, R_j) = \frac{1}{3} \bigg\{ [ H_1 (R_i) - H_1 (R_j) ]^2 + [
H_2 (R_i) - H_2 (R_j) ]^2 + [ M (R_i) - M (R_j) ]^2 \bigg\}
\]</span></p>
<p>This correlation function is actually an average value for the three
amino acid properties: hydrophobicity value, hydrophilicity value and
side chain mass. Therefore, we can extend this definition of correlation
functions for one amino acid property or for a set of <span class="math inline">\(n\)</span> amino acid properties.</p>
<p>For one amino acid property, the correlation can be defined as:</p>
<p><span class="math display">\[
\Theta (R_i, R_j) = [H_1 (R_i) - H_1(R_j)]^2
\]</span></p>
<p>where <span class="math inline">\(H (R_i)\)</span> is the amino acid
property of amino acid <span class="math inline">\(R_i\)</span> after
standardization.</p>
<p>For a set of n amino acid properties, it can be defined as: where
<span class="math inline">\(H_k (R_i)\)</span> is the <span class="math inline">\(k\)</span>-th property in the amino acid property
set for amino acid <span class="math inline">\(R_i\)</span>.</p>
<p><span class="math display">\[
\Theta (R_i, R_j) = \frac{1}{n} \sum_{k=1}^{n} [H_k (R_i) - H_k (R_j)]^2
\]</span></p>
<p>where <span class="math inline">\(H_k(R_i)\)</span> is the <span class="math inline">\(k\)</span>-th property in the amino acid property
set for amino acid <span class="math inline">\(R_i\)</span>.</p>
<p>A set of descriptors named sequence order-correlated factors are
defined as:</p>
<p><span class="math display">\[\begin{align*}
\theta_1  &amp; = \frac{1}{N-1} \sum_{i=1}^{N-1} \Theta (R_i, R_{i+1})\\
\theta_2  &amp; = \frac{1}{N-2} \sum_{i=1}^{N-2} \Theta (R_i, R_{i+2})\\
\theta_3  &amp; = \frac{1}{N-3} \sum_{i=1}^{N-3} \Theta (R_i, R_{i+3})\\
          &amp; \ldots \\
\theta_\lambda  &amp; = \frac{1}{N-\lambda} \sum_{i=1}^{N-\lambda}
\Theta (R_i, R_{i+\lambda})
\end{align*}\]</span></p>
<p><span class="math inline">\(\lambda\)</span> (<span class="math inline">\(\lambda &lt; L\)</span>) is a parameter to be
specified. Let <span class="math inline">\(f_i\)</span> be the
normalized occurrence frequency of the 20 amino acids in the protein
sequence, a set of <span class="math inline">\(20 + \lambda\)</span>
descriptors called the pseudo-amino acid composition for a protein
sequence can be defines as:</p>
<p><span class="math display">\[
X_c = \frac{f_c}{\sum_{r=1}^{20} f_r + w \sum_{j=1}^{\lambda} \theta_j}
\quad (1 &lt; c &lt; 20)
\]</span></p>
<p><span class="math display">\[
X_c = \frac{w \theta_{c-20}}{\sum_{r=1}^{20} f_r + w
\sum_{j=1}^{\lambda} \theta_j} \quad (21 &lt; c &lt; 20 + \lambda)
\]</span></p>
<p>where <span class="math inline">\(w\)</span> is the weighting factor
for the sequence-order effect and is set to <span class="math inline">\(w = 0.05\)</span> in protr as suggested by
Kuo-Chen Chou.</p>
<p>With <code><a href="../reference/extractPAAC.html">extractPAAC()</a></code>, we can compute the PseAAC
descriptors directly:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extractPAAC.html">extractPAAC</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt;         Xc1.A         Xc1.R         Xc1.N         Xc1.D         Xc1.C </span></span>
<span><span class="co">#&gt;    9.07025432   10.07806035    5.54293319    7.30659376    9.57415734 </span></span>
<span><span class="co">#&gt;         Xc1.E         Xc1.Q         Xc1.G         Xc1.H         Xc1.I </span></span>
<span><span class="co">#&gt;    6.80269074    6.80269074   11.58976941    4.28317565    5.03903018 </span></span>
<span><span class="co">#&gt;         Xc1.L         Xc1.K         Xc1.M         Xc1.F         Xc1.P </span></span>
<span><span class="co">#&gt;   10.83391488    5.54293319    1.76366056    4.53512716    7.55854527 </span></span>
<span><span class="co">#&gt;         Xc1.S         Xc1.T         Xc1.W         Xc1.Y         Xc1.V </span></span>
<span><span class="co">#&gt;   12.59757544    6.29878772    3.27536961    6.04683621    7.05464225 </span></span>
<span><span class="co">#&gt;  Xc2.lambda.1  Xc2.lambda.2  Xc2.lambda.3  Xc2.lambda.4  Xc2.lambda.5 </span></span>
<span><span class="co">#&gt;    0.02514092    0.02500357    0.02527773    0.02553159    0.02445265 </span></span>
<span><span class="co">#&gt;  Xc2.lambda.6  Xc2.lambda.7  Xc2.lambda.8  Xc2.lambda.9 Xc2.lambda.10 </span></span>
<span><span class="co">#&gt;    0.02561910    0.02486131    0.02506656    0.02553952    0.02437663 </span></span>
<span><span class="co">#&gt; Xc2.lambda.11 Xc2.lambda.12 Xc2.lambda.13 Xc2.lambda.14 Xc2.lambda.15 </span></span>
<span><span class="co">#&gt;    0.02491262    0.02533803    0.02351915    0.02479912    0.02548431 </span></span>
<span><span class="co">#&gt; Xc2.lambda.16 Xc2.lambda.17 Xc2.lambda.18 Xc2.lambda.19 Xc2.lambda.20 </span></span>
<span><span class="co">#&gt;    0.02478210    0.02513770    0.02457224    0.02543046    0.02500889 </span></span>
<span><span class="co">#&gt; Xc2.lambda.21 Xc2.lambda.22 Xc2.lambda.23 Xc2.lambda.24 Xc2.lambda.25 </span></span>
<span><span class="co">#&gt;    0.02476967    0.02342389    0.02431684    0.02610300    0.02626722 </span></span>
<span><span class="co">#&gt; Xc2.lambda.26 Xc2.lambda.27 Xc2.lambda.28 Xc2.lambda.29 Xc2.lambda.30 </span></span>
<span><span class="co">#&gt;    0.02457082    0.02343049    0.02588823    0.02490463    0.02451951</span></span></code></pre></div>
<p>The <code><a href="../reference/extractPAAC.html">extractPAAC()</a></code> fucntion also provides the additional
arguments <code>props</code> and <code>customprops</code>, which are
similar to those arguments for Moreau-Broto/Moran/Geary autocorrelation
descriptors. For their minor differences, please see
<code>?extracPAAC</code>. Users can specify the lambda parameter and the
weighting factor with arguments <code>lambda</code> and
<code>w</code>.</p>
<p><strong>Note</strong>: In the work of Kuo-Chen Chou, the definition
for â€œnormalized occurrence frequencyâ€ was not given. In this work, we
define it as the occurrence frequency of amino acid in the sequence
normalized to 100% and hence our calculated values are not the same as
values by them.</p>
</div>
<div class="section level3">
<h3 id="amphiphilic-pseudo-amino-acid-composition-apseaac">Amphiphilic pseudo-amino acid composition (APseAAC)<a class="anchor" aria-label="anchor" href="#amphiphilic-pseudo-amino-acid-composition-apseaac"></a>
</h3>
<p>Amphiphilic Pseudo-Amino Acid Composition (APseAAC) was proposed in
<span class="citation">Chou (2001)</span>. APseAAC is also recognized as
the <em>type 2 pseudo-amino acid composition</em>. The definitions of
these qualities are similar to the PAAC descriptors. From <span class="math inline">\(H_1 (i)\)</span> and <span class="math inline">\(H_2 (j)\)</span> defined before, the
hydrophobicity and hydrophilicity correlation functions are defined
as:</p>
<p><span class="math display">\[\begin{align*}
H_{i, j}^1  &amp; = H_1 (i) H_1 (j)\\
H_{i, j}^2  &amp; = H_2 (i) H_2 (j)
\end{align*}\]</span></p>
<p>From these qualities, sequence order factors can be defines as:</p>
<p><span class="math display">\[\begin{align*}
\tau_1  &amp; = \frac{1}{N-1} \sum_{i=1}^{N-1} H_{i, i+1}^1\\
\tau_2  &amp; = \frac{1}{N-1} \sum_{i=1}^{N-1} H_{i, i+1}^2\\
\tau_3  &amp; = \frac{1}{N-2} \sum_{i=1}^{N-2} H_{i, i+2}^1\\
\tau_4  &amp; = \frac{1}{N-2} \sum_{i=1}^{N-2} H_{i, i+2}^2\\
        &amp; \ldots \\
\tau_{2 \lambda - 1} &amp; = \frac{1}{N-\lambda} \sum_{i=1}^{N-\lambda}
H_{i, i+\lambda}^1\\
\tau_{2 \lambda}     &amp; = \frac{1}{N-\lambda} \sum_{i=1}^{N-\lambda}
H_{i, i+\lambda}^2
\end{align*}\]</span></p>
<img src="figures/APAAC.png" width="100%" class="img-responsive" align="center"><p align="left">
Figure 7: A schematic diagram to show
(<strong>a1</strong>/<strong>a2</strong>) the first-rank,
(<strong>b1</strong>/<strong>b2</strong>) the second-rank and
(<strong>c1</strong>/<strong>c2</strong>) the third-rank
sequence-order-coupling mode along a protein sequence through a
hydrophobicity/hydrophilicity correlation function, where <span class="math inline">\(H_{i, j}^1\)</span> and <span class="math inline">\(H_{i, j}^2\)</span> are given by Equation (3).
Panel (a1/a2) reflects the coupling mode between all the most contiguous
residues, panel (b1/b2) that between all the second-most contiguous
residues and panel (c1/c2) that between all the third-most contiguous
residues. This figure is from <span class="citation">Chou (2005)</span>.
</p>
<p>Then a set of descriptors called <em>Amphiphilic Pseudo-Amino Acid
Composition</em> (<em>APseAAC</em>) are defined as:</p>
<p><span class="math display">\[
P_c = \frac{f_c}{\sum_{r=1}^{20} f_r + w \sum_{j=1}^{2 \lambda} \tau_j}
\quad (1 &lt; c &lt; 20)
\]</span></p>
<p><span class="math display">\[
P_c = \frac{w \tau_u}{\sum_{r=1}^{20} f_r + w \sum_{j=1}^{2 \lambda}
\tau_j} \quad (21 &lt; u &lt; 20 + 2 \lambda)
\]</span></p>
<p>where <span class="math inline">\(w\)</span> is the weighting factor.
Its default value is set to <span class="math inline">\(w = 0.5\)</span>
in protr.</p>
<p>A minimal example for <code>extracAPAAC()</code> is:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extractAPAAC.html">extractAPAAC</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt;                 Pc1.A                 Pc1.R                 Pc1.N </span></span>
<span><span class="co">#&gt;          3.537412e+01          3.930458e+01          2.161752e+01 </span></span>
<span><span class="co">#&gt;                 Pc1.D                 Pc1.C                 Pc1.E </span></span>
<span><span class="co">#&gt;          2.849582e+01          3.733935e+01          2.653059e+01 </span></span>
<span><span class="co">#&gt;                 Pc1.Q                 Pc1.G                 Pc1.H </span></span>
<span><span class="co">#&gt;          2.653059e+01          4.520027e+01          1.670445e+01 </span></span>
<span><span class="co">#&gt;                 Pc1.I                 Pc1.L                 Pc1.K </span></span>
<span><span class="co">#&gt;          1.965229e+01          4.225242e+01          2.161752e+01 </span></span>
<span><span class="co">#&gt;                 Pc1.M                 Pc1.F                 Pc1.P </span></span>
<span><span class="co">#&gt;          6.878302e+00          1.768706e+01          2.947844e+01 </span></span>
<span><span class="co">#&gt;                 Pc1.S                 Pc1.T                 Pc1.W </span></span>
<span><span class="co">#&gt;          4.913073e+01          2.456536e+01          1.277399e+01 </span></span>
<span><span class="co">#&gt;                 Pc1.Y                 Pc1.V  Pc2.Hydrophobicity.1 </span></span>
<span><span class="co">#&gt;          2.358275e+01          2.751321e+01          2.196320e-04 </span></span>
<span><span class="co">#&gt;  Pc2.Hydrophilicity.1  Pc2.Hydrophobicity.2  Pc2.Hydrophilicity.2 </span></span>
<span><span class="co">#&gt;          1.025766e-03         -3.088876e-04         -1.834385e-04 </span></span>
<span><span class="co">#&gt;  Pc2.Hydrophobicity.3  Pc2.Hydrophilicity.3  Pc2.Hydrophobicity.4 </span></span>
<span><span class="co">#&gt;          1.174146e-03          7.400156e-04         -1.105715e-03 </span></span>
<span><span class="co">#&gt;  Pc2.Hydrophilicity.4  Pc2.Hydrophobicity.5  Pc2.Hydrophilicity.5 </span></span>
<span><span class="co">#&gt;         -4.493680e-04          1.766358e-03          1.471212e-03 </span></span>
<span><span class="co">#&gt;  Pc2.Hydrophobicity.6  Pc2.Hydrophilicity.6  Pc2.Hydrophobicity.7 </span></span>
<span><span class="co">#&gt;         -1.441572e-03         -4.913600e-03         -1.678053e-05 </span></span>
<span><span class="co">#&gt;  Pc2.Hydrophilicity.7  Pc2.Hydrophobicity.8  Pc2.Hydrophilicity.8 </span></span>
<span><span class="co">#&gt;          7.312356e-04         -1.885399e-03         -1.928708e-03 </span></span>
<span><span class="co">#&gt;  Pc2.Hydrophobicity.9  Pc2.Hydrophilicity.9 Pc2.Hydrophobicity.10 </span></span>
<span><span class="co">#&gt;         -2.931177e-03         -1.555660e-03          2.916597e-03 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophilicity.10 Pc2.Hydrophobicity.11 Pc2.Hydrophilicity.11 </span></span>
<span><span class="co">#&gt;          3.602591e-03          1.055082e-04          8.697920e-04 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophobicity.12 Pc2.Hydrophilicity.12 Pc2.Hydrophobicity.13 </span></span>
<span><span class="co">#&gt;         -9.276413e-04         -2.001384e-03          1.705044e-03 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophilicity.13 Pc2.Hydrophobicity.14 Pc2.Hydrophilicity.14 </span></span>
<span><span class="co">#&gt;          4.364007e-03          7.883453e-04         -9.441693e-04 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophobicity.15 Pc2.Hydrophilicity.15 Pc2.Hydrophobicity.16 </span></span>
<span><span class="co">#&gt;         -3.133437e-04         -3.599332e-03          3.689079e-05 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophilicity.16 Pc2.Hydrophobicity.17 Pc2.Hydrophilicity.17 </span></span>
<span><span class="co">#&gt;          2.483867e-03          4.832798e-04          2.465788e-03 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophobicity.18 Pc2.Hydrophilicity.18 Pc2.Hydrophobicity.19 </span></span>
<span><span class="co">#&gt;         -3.142728e-04          2.021961e-03          6.421283e-05 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophilicity.19 Pc2.Hydrophobicity.20 Pc2.Hydrophilicity.20 </span></span>
<span><span class="co">#&gt;         -8.896690e-04         -2.986886e-04          9.304039e-04 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophobicity.21 Pc2.Hydrophilicity.21 Pc2.Hydrophobicity.22 </span></span>
<span><span class="co">#&gt;         -6.777458e-04          1.646818e-03          3.193506e-03 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophilicity.22 Pc2.Hydrophobicity.23 Pc2.Hydrophilicity.23 </span></span>
<span><span class="co">#&gt;          3.270656e-03          2.533569e-03          2.478252e-03 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophobicity.24 Pc2.Hydrophilicity.24 Pc2.Hydrophobicity.25 </span></span>
<span><span class="co">#&gt;         -2.489106e-03         -1.031008e-03         -3.992322e-03 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophilicity.25 Pc2.Hydrophobicity.26 Pc2.Hydrophilicity.26 </span></span>
<span><span class="co">#&gt;         -2.596060e-03          8.690771e-04         -1.221378e-03 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophobicity.27 Pc2.Hydrophilicity.27 Pc2.Hydrophobicity.28 </span></span>
<span><span class="co">#&gt;          5.208649e-03          4.617400e-03         -1.088584e-03 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophilicity.28 Pc2.Hydrophobicity.29 Pc2.Hydrophilicity.29 </span></span>
<span><span class="co">#&gt;         -2.512263e-03          1.387641e-03          2.060890e-03 </span></span>
<span><span class="co">#&gt; Pc2.Hydrophobicity.30 Pc2.Hydrophilicity.30 </span></span>
<span><span class="co">#&gt;          3.177340e-04          1.451909e-03</span></span></code></pre></div>
<p>This function has the same arguments as
<code><a href="../reference/extractPAAC.html">extractPAAC()</a></code>.</p>
</div>
<div class="section level3">
<h3 id="profile-based-descriptors">Profile-based descriptors<a class="anchor" aria-label="anchor" href="#profile-based-descriptors"></a>
</h3>
<p>The profile-based descriptors for protein sequences are available in
the protr package. The feature vectors of profile-based methods were
based on the PSSM by running PSI-BLAST, and often show good performance.
See <span class="citation">Ye <em>et al.</em> (2011)</span> and <span class="citation">Rangwala and Karypis (2005)</span> for details. The
functions <code><a href="../reference/extractPSSM.html">extractPSSM()</a></code>, <code><a href="../reference/extractPSSMAcc.html">extractPSSMAcc()</a></code> and
<code><a href="../reference/extractPSSMFeature.html">extractPSSMFeature()</a></code> are used to generate these
descriptors. Users need to install the NCBI-BLAST+ software package
first to make the functions fully functional.</p>
</div>
</div>
<div class="section level2">
<h2 id="proteochemometric-modeling-descriptors">Proteochemometric modeling descriptors<a class="anchor" aria-label="anchor" href="#proteochemometric-modeling-descriptors"></a>
</h2>
<p>Proteochemometric (PCM) modeling utilizes statistical modeling
techniques to model ligand-target interaction space. The below
descriptors implemented in protr are extensively used in
Proteochemometric modeling.</p>
<ul>
<li>
<p>Scales-based descriptors derived by Principal Components
Analysis</p>
<ul>
<li>Scales-based descriptors derived by Amino Acid Properties from
AAindex (Protein Fingerprint)</li>
<li>Scales-based descriptors derived by 20+ classes of 2D and 3D
molecular descriptors (Topological, WHIM, VHSE, etc.)</li>
</ul>
</li>
<li><p>Scales-based descriptors derived by Factor Analysis</p></li>
<li><p>Scales-based descriptors derived by Multidimensional
Scaling</p></li>
<li><p>BLOSUM and PAM matrix-derived descriptors</p></li>
</ul>
<p>Note that each of the scales-based descriptor functions are freely to
combine with the more than 20 classes of 2D and 3D molecular descriptors
to construct highly customized scales-based descriptors. Of course,
these functions are designed to be flexible enough that users can
provide totally self-defined property matrices to construct scales-based
descriptors.</p>
<p>For example, to compute the â€œtopological scalesâ€ derived by PCA
(using the first 5 principal components), one can use
<code><a href="../reference/extractDescScales.html">extractDescScales()</a></code>:</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"protseq/P00750.fasta"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"protr"</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">descscales</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractDescScales.html">extractDescScales</a></span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  propmat <span class="op">=</span> <span class="st">"AATopo"</span>,</span>
<span>  index <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">37</span><span class="op">:</span><span class="fl">41</span>, <span class="fl">43</span><span class="op">:</span><span class="fl">47</span><span class="op">)</span>,</span>
<span>  pc <span class="op">=</span> <span class="fl">5</span>, lag <span class="op">=</span> <span class="fl">7</span>, silent <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Summary of the first 5 principal components: </span></span>
<span><span class="co">#&gt;                             PC1      PC2       PC3       PC4        PC5</span></span>
<span><span class="co">#&gt; Standard deviation     2.581537 1.754133 0.4621854 0.1918666 0.08972087</span></span>
<span><span class="co">#&gt; Proportion of Variance 0.666430 0.307700 0.0213600 0.0036800 0.00080000</span></span>
<span><span class="co">#&gt; Cumulative Proportion  0.666430 0.974130 0.9954900 0.9991700 0.99998000</span></span></code></pre></div>
<p>the argument <code>propmat</code> involkes the <code>AATopo</code>
dataset shipped with the protr package, and the argument
<code>index</code> selects the 37 to 41 and the 43 to 47 columns
(molecular descriptors) in the <code>AATopo</code> dataset to use, the
parameter <code>lag</code> was set for the Auto Cross Covariance (ACC)
for generating scales-based descriptors of the same length. At last, we
printed the summary of the first 5 principal components (standard
deviation, proportion of variance, cumulative proportion of
variance).</p>
<p>The result is a length 175 named vector, which is consistent with the
descriptors before:</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">descscales</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 175</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">descscales</span>, <span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">#&gt;     scl1.lag1     scl2.lag1     scl3.lag1     scl4.lag1     scl5.lag1 </span></span>
<span><span class="co">#&gt; -2.645644e-01 -1.717847e-02  1.975438e-02 -7.930659e-05 -3.710597e-05 </span></span>
<span><span class="co">#&gt;     scl1.lag2     scl2.lag2     scl3.lag2     scl4.lag2     scl5.lag2 </span></span>
<span><span class="co">#&gt;  3.548612e-01  1.343712e-01  5.699395e-03 -5.489472e-04 -6.364577e-05 </span></span>
<span><span class="co">#&gt;     scl1.lag3     scl2.lag3     scl3.lag3     scl4.lag3     scl5.lag3 </span></span>
<span><span class="co">#&gt;  2.011431e-02 -9.211136e-02 -1.461755e-03  6.747801e-04  2.386782e-04</span></span></code></pre></div>
<p>For another example, to compute the descriptors derived by the
BLOSUM62 matrix and use the first 5 scales, one can use:</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>  <span class="st">"protseq/P00750.fasta"</span>,</span>
<span>  package <span class="op">=</span> <span class="st">"protr"</span></span>
<span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="va">blosum</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extractBLOSUM.html">extractBLOSUM</a></span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  submat <span class="op">=</span> <span class="st">"AABLOSUM62"</span>,</span>
<span>  k <span class="op">=</span> <span class="fl">5</span>, lag <span class="op">=</span> <span class="fl">7</span>, scale <span class="op">=</span> <span class="cn">TRUE</span>, silent <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Relative importance of all the possible 20 scales: </span></span>
<span><span class="co">#&gt;  [1]  1.204960e+01  7.982007e+00  6.254364e+00  4.533706e+00  4.326286e+00</span></span>
<span><span class="co">#&gt;  [6]  3.850579e+00  3.752197e+00  3.538207e+00  3.139155e+00  2.546405e+00</span></span>
<span><span class="co">#&gt; [11]  2.373286e+00  1.666259e+00  1.553126e+00  1.263685e+00  1.024699e+00</span></span>
<span><span class="co">#&gt; [16]  9.630187e-01  9.225759e-01  7.221636e-01  1.020085e-01 -5.354392e-18</span></span></code></pre></div>
<p>The result is a length 175 named vector:</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">blosum</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 175</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">blosum</span>, <span class="fl">15</span><span class="op">)</span></span>
<span><span class="co">#&gt;     scl1.lag1     scl2.lag1     scl3.lag1     scl4.lag1     scl5.lag1 </span></span>
<span><span class="co">#&gt;  0.0042370555 -0.0021502057  0.0005993291  0.0006456375  0.0014849592 </span></span>
<span><span class="co">#&gt;     scl1.lag2     scl2.lag2     scl3.lag2     scl4.lag2     scl5.lag2 </span></span>
<span><span class="co">#&gt; -0.0014919096  0.0032873726  0.0011734162 -0.0021758536 -0.0018127568 </span></span>
<span><span class="co">#&gt;     scl1.lag3     scl2.lag3     scl3.lag3     scl4.lag3     scl5.lag3 </span></span>
<span><span class="co">#&gt; -0.0029413528  0.0001494193  0.0003298806 -0.0017877430 -0.0051044133</span></span></code></pre></div>
<p><strong>Dealing with gaps.</strong> In proteochemometrics, (sequence
alignment) gaps can be very useful, since a gap in a certain position
contains information. The protr package has built-in support for such
gaps. We deal with the gaps by using a dummy descriptor to code for the
21st type of amino acid. The function <code><a href="../reference/extractScalesGap.html">extractScalesGap()</a></code>
and <code><a href="../reference/extractProtFPGap.html">extractProtFPGap()</a></code> can be used to deal with such gaps.
See <code><a href="../reference/extractScalesGap.html">?extractScalesGap</a></code> and <code><a href="../reference/extractProtFPGap.html">?extractProtFPGap</a></code>
for details.</p>
</div>
<div class="section level2">
<h2 id="similarity-calculation-by-sequence-alignment">Similarity calculation by sequence alignment<a class="anchor" aria-label="anchor" href="#similarity-calculation-by-sequence-alignment"></a>
</h2>
<p>Similarity computation derived by local or global protein sequence
alignment between a list of protein sequences is of great need in
protein research. However, this type of pairwise similarity computation
often computationally intensive, especially when there exists many
protein sequences. Luckily, this process is also highly parallelizable,
the protr package integrates the function of parallelized similarity
computation derived by local or global protein sequence alignment
between a list of protein sequences.</p>
<p>The function <code><a href="../reference/twoSeqSim.html">twoSeqSim()</a></code> calculates the alignment result
between two protein sequences. The function <code><a href="../reference/parSeqSim.html">parSeqSim()</a></code>
calculates the pairwise similarity calculation with a list of protein
sequences in parallel:</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">s1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"protseq/P00750.fasta"</span>, package <span class="op">=</span> <span class="st">"protr"</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">s2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"protseq/P08218.fasta"</span>, package <span class="op">=</span> <span class="st">"protr"</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">s3</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"protseq/P10323.fasta"</span>, package <span class="op">=</span> <span class="st">"protr"</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">s4</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"protseq/P20160.fasta"</span>, package <span class="op">=</span> <span class="st">"protr"</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">s5</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"protseq/Q9NZP8.fasta"</span>, package <span class="op">=</span> <span class="st">"protr"</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">plist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">s1</span>, <span class="va">s2</span>, <span class="va">s3</span>, <span class="va">s4</span>, <span class="va">s5</span><span class="op">)</span></span>
<span><span class="va">psimmat</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parSeqSim.html">parSeqSim</a></span><span class="op">(</span><span class="va">plist</span>, cores <span class="op">=</span> <span class="fl">4</span>, type <span class="op">=</span> <span class="st">"local"</span>, submat <span class="op">=</span> <span class="st">"BLOSUM62"</span><span class="op">)</span></span>
<span><span class="va">psimmat</span></span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##            [,1]       [,2]       [,3]       [,4]       [,5]</span></span>
<span><span class="co">## [1,] 1.00000000 0.11825938 0.10236985 0.04921696 0.03943488</span></span>
<span><span class="co">## [2,] 0.11825938 1.00000000 0.18858241 0.12124217 0.06391103</span></span>
<span><span class="co">## [3,] 0.10236985 0.18858241 1.00000000 0.05819984 0.06175942</span></span>
<span><span class="co">## [4,] 0.04921696 0.12124217 0.05819984 1.00000000 0.05714638</span></span>
<span><span class="co">## [5,] 0.03943488 0.06391103 0.06175942 0.05714638 1.00000000</span></span></code></pre></div>
<p>Note that for a small number of proteins, calculating their pairwise
similarity scores derived by sequence alignment in parallel may not
significantly reduce the overall computation time, since each of the
task only requires a relatively small time to finish, thus,
computational overheads may exist and affect the performance. In our
benchmarks, we used about 1,000 protein sequences on 64 CPU cores, and
observed significant performance improvement comparing to the sequential
computation.</p>
<div class="section level3">
<h3 id="scaling-up-for-a-large-number-of-sequences">Scaling up for a large number of sequences<a class="anchor" aria-label="anchor" href="#scaling-up-for-a-large-number-of-sequences"></a>
</h3>
<p><strong><code>batches</code></strong>. A useful argument for reducing
the memory usage is <code>batches</code>. It splits the similarity
computations into smaller batches. A larger batch number reduces the
memory need for each parallel process, but also adds the overhead of
forking new processes. This is a trade-off between time (CPU) and space
(memory). Also, use <code>verbose = TRUE</code> to print the computation
progress.</p>
<p><strong><code><a href="../reference/parSeqSimDisk.html">parSeqSimDisk()</a></code></strong>. For similarity
computations over an even larger number of protein sequences, please try
<code><a href="../reference/parSeqSimDisk.html">parSeqSimDisk()</a></code> and set <code>batches</code> to a large
number. Unlike the in-memory version, this disk-based function caches
the partial results in each batch to the hard drive and merges all the
results together in the end, which further reduces the memory usage.</p>
<p><strong>Example</strong>. Letâ€™s assume we have 20,000 sequences. By
setting <code>cores = 64</code> and <code>batches = 10000</code> in
<code><a href="../reference/parSeqSimDisk.html">parSeqSimDisk()</a></code>, the workload will be around
<code>20000^2/2/64/10000 = 312.5</code> alignments per batch per core.
This could give you a relatively quick feedback about if the memory size
is going to be sufficient, and how long the computation will take in
total, after the first few batches.</p>
</div>
</div>
<div class="section level2">
<h2 id="similarity-calculation-by-go-semantic-similarity-measures">Similarity calculation by GO semantic similarity measures<a class="anchor" aria-label="anchor" href="#similarity-calculation-by-go-semantic-similarity-measures"></a>
</h2>
<p>The protr package also integrates the function of similarity score
computation derived by Gene Ontology (GO) semantic similarity measures
between a list of GO terms or Entrez Gene IDs.</p>
<p>The function <code><a href="../reference/twoGOSim.html">twoGOSim()</a></code> calculates the similarity
derived by GO-terms semantic similarity measures between two GO terms /
Entrez Gene IDs, and the function <code><a href="../reference/parGOSim.html">parGOSim()</a></code> calculates the
pairwise similarity with a list of GO terms / Entrez Gene IDs:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By GO Terms</span></span>
<span><span class="va">go1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"GO:0005215"</span>, <span class="st">"GO:0005488"</span>, <span class="st">"GO:0005515"</span>,</span>
<span>  <span class="st">"GO:0005625"</span>, <span class="st">"GO:0005802"</span>, <span class="st">"GO:0005905"</span></span>
<span><span class="op">)</span> <span class="co"># AP4B1</span></span>
<span><span class="va">go2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"GO:0005515"</span>, <span class="st">"GO:0005634"</span>, <span class="st">"GO:0005681"</span>,</span>
<span>  <span class="st">"GO:0008380"</span>, <span class="st">"GO:0031202"</span></span>
<span><span class="op">)</span> <span class="co"># BCAS2</span></span>
<span><span class="va">go3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"GO:0003735"</span>, <span class="st">"GO:0005622"</span>, <span class="st">"GO:0005840"</span>,</span>
<span>  <span class="st">"GO:0006412"</span></span>
<span><span class="op">)</span> <span class="co"># PDE4DIP</span></span>
<span><span class="va">golist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">go1</span>, <span class="va">go2</span>, <span class="va">go3</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/parGOSim.html">parGOSim</a></span><span class="op">(</span><span class="va">golist</span>, type <span class="op">=</span> <span class="st">"go"</span>, ont <span class="op">=</span> <span class="st">"CC"</span>, measure <span class="op">=</span> <span class="st">"Wang"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##       [,1]  [,2] [,3]</span></span>
<span><span class="co">## [1,] 1.000 0.344 0.17</span></span>
<span><span class="co">## [2,] 0.344 1.000 0.24</span></span>
<span><span class="co">## [3,] 0.170 0.240 1.00</span></span></code></pre></div>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># By Entrez gene id</span></span>
<span><span class="va">genelist</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"150"</span>, <span class="st">"151"</span>, <span class="st">"152"</span>, <span class="st">"1814"</span>, <span class="st">"1815"</span>, <span class="st">"1816"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/parGOSim.html">parGOSim</a></span><span class="op">(</span><span class="va">genelist</span>, type <span class="op">=</span> <span class="st">"gene"</span>, ont <span class="op">=</span> <span class="st">"BP"</span>, measure <span class="op">=</span> <span class="st">"Wang"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">##        150   151   152  1814  1815  1816</span></span>
<span><span class="co">## 150  1.000 0.702 0.725 0.496 0.570 0.455</span></span>
<span><span class="co">## 151  0.702 1.000 0.980 0.456 0.507 0.551</span></span>
<span><span class="co">## 152  0.725 0.980 1.000 0.460 0.511 0.538</span></span>
<span><span class="co">## 1814 0.496 0.456 0.460 1.000 0.687 0.473</span></span>
<span><span class="co">## 1815 0.570 0.507 0.511 0.687 1.000 0.505</span></span>
<span><span class="co">## 1816 0.455 0.551 0.538 0.473 0.505 1.000</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="miscellaneous-tools">Miscellaneous tools<a class="anchor" aria-label="anchor" href="#miscellaneous-tools"></a>
</h2>
<p>In this section, we will briefly introduce some useful tools provided
by the protr package.</p>
<div class="section level3">
<h3 id="retrieve-protein-sequences-from-uniprot">Retrieve protein sequences from UniProt<a class="anchor" aria-label="anchor" href="#retrieve-protein-sequences-from-uniprot"></a>
</h3>
<p>This function <code><a href="../reference/getUniProt.html">getUniProt()</a></code> gets protein sequences from
uniprot.org by protein ID(s). The input <code>ID</code> is a character
vector specifying the protein ID(s). The returned sequences are stored
in a list:</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"P00750"</span>, <span class="st">"P00751"</span>, <span class="st">"P00752"</span><span class="op">)</span></span>
<span><span class="va">prots</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/getUniProt.html">getUniProt</a></span><span class="op">(</span><span class="va">ids</span><span class="op">)</span></span>
<span><span class="va">prots</span></span></code></pre></div>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [1] "MDAMKRGLCCVLLLCGAVFVSPSQEIHARFRRGARSYQVICRDEKTQMIYQQHQSWLRPVLRSNRVEYCWCN</span></span>
<span><span class="co">## SGRAQCHSVPVKSCSEPRCFNGGTCQQALYFSDFVCQCPEGFAGKCCEIDTRATCYEDQGISYRGTWSTAESGAECT</span></span>
<span><span class="co">## NWNSSALAQKPYSGRRPDAIRLGLGNHNYCRNPDRDSKPWCYVFKAGKYSSEFCSTPACSEGNSDCYFGNGSAYRGT</span></span>
<span><span class="co">## HSLTESGASCLPWNSMILIGKVYTAQNPSAQALGLGKHNYCRNPDGDAKPWCHVLKNRRLTWEYCDVPSCSTCGLRQ</span></span>
<span><span class="co">## YSQPQFRIKGGLFADIASHPWQAAIFAKHRRSPGERFLCGGILISSCWILSAAHCFQERFPPHHLTVILGRTYRVVP</span></span>
<span><span class="co">## GEEEQKFEVEKYIVHKEFDDDTYDNDIALLQLKSDSSRCAQESSVVRTVCLPPADLQLPDWTECELSGYGKHEALSP</span></span>
<span><span class="co">## FYSERLKEAHVRLYPSSRCTSQHLLNRTVTDNMLCAGDTRSGGPQANLHDACQGDSGGPLVCLNDGRMTLVGIISWG</span></span>
<span><span class="co">## LGCGQKDVPGVYTKVTNYLDWIRDNMRP"</span></span>
<span><span class="co">##</span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [1] "MGSNLSPQLCLMPFILGLLSGGVTTTPWSLARPQGSCSLEGVEIKGGSFRLLQEGQALEYVCPSGFYPYPVQ</span></span>
<span><span class="co">## TRTCRSTGSWSTLKTQDQKTVRKAECRAIHCPRPHDFENGEYWPRSPYYNVSDEISFHCYDGYTLRGSANRTCQVNG</span></span>
<span><span class="co">## RWSGQTAICDNGAGYCSNPGIPIGTRKVGSQYRLEDSVTYHCSRGLTLRGSQRRTCQEGGSWSGTEPSCQDSFMYDT</span></span>
<span><span class="co">## PQEVAEAFLSSLTETIEGVDAEDGHGPGEQQKRKIVLDPSGSMNIYLVLDGSDSIGASNFTGAKKCLVNLIEKVASY</span></span>
<span><span class="co">## GVKPRYGLVTYATYPKIWVKVSEADSSNADWVTKQLNEINYEDHKLKSGTNTKKALQAVYSMMSWPDDVPPEGWNRT</span></span>
<span><span class="co">## RHVIILMTDGLHNMGGDPITVIDEIRDLLYIGKDRKNPREDYLDVYVFGVGPLVNQVNINALASKKDNEQHVFKVKD</span></span>
<span><span class="co">## MENLEDVFYQMIDESQSLSLCGMVWEHRKGTDYHKQPWQAKISVIRPSKGHESCMGAVVSEYFVLTAAHCFTVDDKE</span></span>
<span><span class="co">## HSIKVSVGGEKRDLEIEVVLFHPNYNINGKKEAGIPEFYDYDVALIKLKNKLKYGQTIRPICLPCTEGTTRALRLPP</span></span>
<span><span class="co">## TTTCQQQKEELLPAQDIKALFVSEEEKKLTRKEVYIKNGDKKGSCERDAQYAPGYDKVKDISEVVTPRFLCTGGVSP</span></span>
<span><span class="co">## YADPNTCRGDSGGPLIVHKRSRFIQVGVISWGVVDVCKNQKRQKQVPAHARDFHINLFQVLPWLKEKLQDEDLGFL"</span></span>
<span><span class="co">##</span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## [1] "APPIQSRIIGGRECEKNSHPWQVAIYHYSSFQCGGVLVNPKWVLTAAHCKNDNYEVWLGRHNLFENENTAQF</span></span>
<span><span class="co">## FGVTADFPHPGFNLSLLKXHTKADGKDYSHDLMLLRLQSPAKITDAVKVLELPTQEPELGSTCEASGWGSIEPGPDB</span></span>
<span><span class="co">## FEFPDEIQCVQLTLLQNTFCABAHPBKVTESMLCAGYLPGGKDTCMGDSGGPLICNGMWQGITSWGHTPCGSANKPS</span></span>
<span><span class="co">## IYTKLIFYLDWINDTITENP"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="read-fasta-files">Read FASTA files<a class="anchor" aria-label="anchor" href="#read-fasta-files"></a>
</h3>
<p>The function <code><a href="../reference/readFASTA.html">readFASTA()</a></code> provides a convenient way to
read protein sequences stored in FASTA format files. See
<code><a href="../reference/readFASTA.html">?readFASTA</a></code> for details. The returned sequences are stored
in a named list, whose components are named with the protein sequencesâ€™
names.</p>
</div>
<div class="section level3">
<h3 id="read-pdb-files">Read PDB files<a class="anchor" aria-label="anchor" href="#read-pdb-files"></a>
</h3>
<p>The Protein Data Bank (PDB) file format is a text file format
describing the three dimensional structures of protein. The function
<code><a href="../reference/readPDB.html">readPDB()</a></code> provides the function to read protein sequences
stored in PDB format files. See <code><a href="../reference/readPDB.html">?readPDB</a></code> for details.</p>
</div>
<div class="section level3">
<h3 id="sanity-check-for-amino-acid-types">Sanity check for amino acid types<a class="anchor" aria-label="anchor" href="#sanity-check-for-amino-acid-types"></a>
</h3>
<p>The <code><a href="../reference/protcheck.html">protcheck()</a></code> function checks if the protein
sequenceâ€™s amino acid types are in the 20 default types, which returns a
<code>TRUE</code> if all the amino acids in the sequence belongs to the
20 default types:</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/readFASTA.html">readFASTA</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"protseq/P00750.fasta"</span>, package <span class="op">=</span> <span class="st">"protr"</span><span class="op">)</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co"># a real sequence</span></span>
<span><span class="fu"><a href="../reference/protcheck.html">protcheck</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co"># an artificial sequence</span></span>
<span><span class="fu"><a href="../reference/protcheck.html">protcheck</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="va">x</span>, <span class="st">"Z"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="remove-gaps-from-sequences">Remove gaps from sequences<a class="anchor" aria-label="anchor" href="#remove-gaps-from-sequences"></a>
</h3>
<p>The <code><a href="../reference/removeGaps.html">removeGaps()</a></code> function allows us to remove/replace
gaps (<code>-</code>) or any â€œirregularâ€ characters from amino acid
sequences, to prepare them for feature extraction or sequence alignment
based similarity computation.</p>
</div>
<div class="section level3">
<h3 id="protein-sequence-partition">Protein sequence partition<a class="anchor" aria-label="anchor" href="#protein-sequence-partition"></a>
</h3>
<p>The <code><a href="../reference/protseg.html">protseg()</a></code> function partitions the protein sequences
to create sliding windows. This is usually required when creating
feature vectors for machine learning tasks. Users can specify a sequence
<code>x</code>, and a character <code>aa</code>, one of the 20 amino
acid types, and a positive integer <code>k</code>, which controls the
window size (half of the window).</p>
<p>This function returns a named list, each component contains one of
the segmentations (a character string), names of the list components are
the positions of the specified amino acid in the sequence. See the
example below:</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/protseg.html">protseg</a></span><span class="op">(</span><span class="va">x</span>, aa <span class="op">=</span> <span class="st">"M"</span>, k <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span><span class="co">#&gt; $`48`</span></span>
<span><span class="co">#&gt; [1] "DEKTQMIYQQH"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`242`</span></span>
<span><span class="co">#&gt; [1] "LPWNSMILIGK"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`490`</span></span>
<span><span class="co">#&gt; [1] "TVTDNMLCAGD"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $`525`</span></span>
<span><span class="co">#&gt; [1] "LNDGRMTLVGI"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="auto-cross-covariance-computation">Auto cross covariance computation<a class="anchor" aria-label="anchor" href="#auto-cross-covariance-computation"></a>
</h3>
<p>Auto Cross Covariance (ACC) is extensively used in the scales-based
descriptors computation, this approach calculates the auto covariance
and auto cross covariance for generating scale-based descriptors of the
same length. Users can write their own scales-based descriptor functions
with the help of <code><a href="../reference/ACC.html">acc()</a></code> function in the protr package.</p>
</div>
<div class="section level3">
<h3 id="pre-computed-2d-and-3d-descriptor-sets-for-the-20-amino-acids">Pre-computed 2D and 3D descriptor sets for the 20 amino acids<a class="anchor" aria-label="anchor" href="#pre-computed-2d-and-3d-descriptor-sets-for-the-20-amino-acids"></a>
</h3>
<p>The protr package ships with more than 20 pre-computed 2D and 3D
descriptor sets for the 20 amino acids to use with the scales-based
descriptors. Please use <code>data(package = "protr")</code> to list all
the datasets included in the protr package.</p>
</div>
<div class="section level3">
<h3 id="blosum-and-pam-matrices-for-the-20-amino-acids">BLOSUM and PAM matrices for the 20 amino acids<a class="anchor" aria-label="anchor" href="#blosum-and-pam-matrices-for-the-20-amino-acids"></a>
</h3>
<p>The BLOSUM and PAM matrices for the 20 amino acids can be used to
calculate BLOSUM and PAM matrix-derived descriptors with function
<code><a href="../reference/extractBLOSUM.html">extractBLOSUM()</a></code>, the datasets are named in
<code>AABLOSUM45</code>, <code>AABLOSUM50</code>,
<code>AABLOSUM62</code>, <code>AABLOSUM80</code>,
<code>AABLOSUM100</code>, <code>AAPAM30</code>, <code>AAPAM40</code>,
<code>AAPAM70</code>, <code>AAPAM120</code>, and
<code>AAPAM250</code>.</p>
</div>
<div class="section level3">
<h3 id="metadata-of-the-20-amino-acids">Metadata of the 20 amino acids<a class="anchor" aria-label="anchor" href="#metadata-of-the-20-amino-acids"></a>
</h3>
<p>As the reference, the <code>AAMetaInfo</code> dataset includes the
meta information of the 20 amino acids used for the 2D and 3D descriptor
calculation in the protr package. This dataset include each amino acidâ€™s
name, one-letter representation, three-letter representation, SMILE
representation, PubChem CID and PubChem link. See
<code>data(AAMetaInfo)</code> for details.</p>
</div>
</div>
<div class="section level2">
<h2 id="protrweb">ProtrWeb<a class="anchor" aria-label="anchor" href="#protrweb"></a>
</h2>
<p>The web service built on protr, namely ProtrWeb, is located at <a href="http://protr.org" class="external-link uri">http://protr.org</a>.</p>
<p>ProtrWeb (Figure 8) does not require any knowledge of programming for
the user. It is a user-friendly web application for computing the
protein sequence descriptors presented in the protr package.</p>
<img src="figures/protrweb.png" width="100%" class="img-responsive" align="center" alt="ProtrWeb"><p align="center">
Figure 8: A screenshot of the web application ProtrWeb
</p>
<p>Source code repository for this Shiny web application: <a href="https://github.com/nanxstats/protrweb" class="external-link uri">https://github.com/nanxstats/protrweb</a>.</p>
</div>
<div class="section level2">
<h2 id="summary">Summary<a class="anchor" aria-label="anchor" href="#summary"></a>
</h2>
<p>The summary of the descriptors in the protr package are listed in
Table 3.</p>
<table class="table">
<caption>Table 3: List of commonly used descriptors in protr</caption>
<thead><tr class="header">
<th align="left">Descriptor Group</th>
<th align="left">Descriptor Name</th>
<th align="left">Descriptor Dimension</th>
<th align="left">Function Name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Amino Acid Composition</td>
<td align="left">Amino Acid Composition</td>
<td align="left">20</td>
<td align="left">extractAAC()</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">Dipeptide Composition</td>
<td align="left">400</td>
<td align="left">extractDC()</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">Tripeptide Composition</td>
<td align="left">8000</td>
<td align="left">extractTC()</td>
</tr>
<tr class="even">
<td align="left">Autocorrelation</td>
<td align="left">Normalized Moreau-Broto Autocorrelation</td>
<td align="left">240<span class="math inline">\(^1\)</span>
</td>
<td align="left">extractMoreauBroto()</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">Moran Autocorrelation</td>
<td align="left">240<span class="math inline">\(^1\)</span>
</td>
<td align="left">extractMoran()</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">Geary Autocorrelation</td>
<td align="left">240<span class="math inline">\(^1\)</span>
</td>
<td align="left">extractGeary()</td>
</tr>
<tr class="odd">
<td align="left">CTD</td>
<td align="left">Composition</td>
<td align="left">21</td>
<td align="left">extractCTDC(), extractCTDCClass()</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">Transition</td>
<td align="left">21</td>
<td align="left">extractCTDT(), extractCTDTClass()</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">Distribution</td>
<td align="left">105</td>
<td align="left">extractCTDD(), extractCTDDClass()</td>
</tr>
<tr class="even">
<td align="left">Conjoint Triad</td>
<td align="left">Conjoint Triad</td>
<td align="left">343</td>
<td align="left">extractCTriad(), extractCTriadClass()</td>
</tr>
<tr class="odd">
<td align="left">Quasi-Sequence-Order</td>
<td align="left">Sequence-Order-Coupling Number</td>
<td align="left">60<span class="math inline">\(^2\)</span>
</td>
<td align="left">extractSOCN()</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">Quasi-Sequence-Order Descriptors</td>
<td align="left">100<span class="math inline">\(^2\)</span>
</td>
<td align="left">extractQSO()</td>
</tr>
<tr class="odd">
<td align="left">Pseudo-Amino Acid Composition</td>
<td align="left">Pseudo-Amino Acid Composition</td>
<td align="left">50<span class="math inline">\(^3\)</span>
</td>
<td align="left">extractPAAC()</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">Amphiphilic Pseudo-Amino Acid Composition</td>
<td align="left">80<span class="math inline">\(^4\)</span>
</td>
<td align="left">extractAPAAC()</td>
</tr>
</tbody>
</table>
<p align="left">
<sup>1</sup> The number depends on the choice of the number of
properties of amino acids and the choice of the maximum values of the
<code>lag</code>. The default is use 8 types of properties and
<code>lag = 30</code>.
</p>
<p align="left">
<sup>2</sup> The number depends on the maximum value of
<code>lag</code>. By default, <code>lag = 30</code>. Two distance
matrices are used, so the descriptor dimension is <span class="math inline">\(30 \times 2 = 60\)</span> and <span class="math inline">\((20 + 30) \times 2 = 100\)</span>.
</p>
<p align="left">
<sup>3</sup> The number depends on the choice of the number of the set
of amino acid properties and the choice of the <span class="math inline">\(\lambda\)</span> value. The default is use 3 types
of properties proposed by Kuo-Chen Chou and <span class="math inline">\(\lambda = 30\)</span>.
</p>
<p align="left">
<sup>4</sup> The number depends on the choice of the <span class="math inline">\(\lambda\)</span> vlaue. The default is <span class="math inline">\(\lambda = 30\)</span>.
</p>
<p>The scales-based PCM descriptors in the protr package are listed in
Table 4.</p>
<table class="table">
<caption>Table 4: List of PCM descriptors in protr</caption>
<thead><tr class="header">
<th align="left">Derived by</th>
<th align="left">Descriptor Class</th>
<th align="left">Function Name</th>
<th></th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Principal Components Analysis</td>
<td align="left">Scales-based descriptors derived by Principal
Components Analysis</td>
<td align="left">extractScales(), extractScalesGap()</td>
<td></td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">Scales-based descriptors derived by amino acid
properties from AAindex (a.k.a. Protein Fingerprint)</td>
<td align="left">extractProtFP(), extractProtFPGap()</td>
<td></td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">Scales-based descriptors derived by 2D and 3D molecular
descriptors (Topological, WHIM, VHSE, etc.)</td>
<td align="left">extractDescScales()</td>
<td></td>
</tr>
<tr class="even">
<td align="left">Factor Analysis</td>
<td align="left">Scales-based descriptors derived by Factor
Analysis</td>
<td align="left">extractFAScales()</td>
<td></td>
</tr>
<tr class="odd">
<td align="left">Multidimensional Scaling</td>
<td align="left">Scales-based descriptors derived by Multidimensional
Scaling</td>
<td align="left">extractMDSScales()</td>
<td></td>
</tr>
<tr class="even">
<td align="left">Substitution Matrix</td>
<td align="left">BLOSUM and PAM matrix-derived descriptors</td>
<td align="left">extractBLOSUM()</td>
<td></td>
</tr>
</tbody>
</table>
<p>The amino acid descriptor sets used by scales-based descriptors
provided by the protr package are listed in Table 5. Note that the
non-informative descriptors (like the descriptors have only one value
across all the 20 amino acids) in these datasets have already been
filtered out.</p>
<table class="table">
<caption>Table 5: List of the pre-calculated descriptor sets of the 20
amino acids in protr</caption>
<thead><tr class="header">
<th align="left">Dataset Name</th>
<th align="left">Descriptor Set Name</th>
<th align="left">Dimensionality</th>
<th align="left">Calculated by</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">AA2DACOR</td>
<td align="left">2D Autocorrelations Descriptors</td>
<td align="left">92</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AA3DMoRSE</td>
<td align="left">3D-MoRSE Descriptors</td>
<td align="left">160</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAACF</td>
<td align="left">Atom-Centred Fragments Descriptors</td>
<td align="left">6</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AABurden</td>
<td align="left">Burden Eigenvalues Descriptors</td>
<td align="left">62</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAConn</td>
<td align="left">Connectivity Indices Descriptors</td>
<td align="left">33</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAConst</td>
<td align="left">Constitutional Descriptors</td>
<td align="left">23</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAEdgeAdj</td>
<td align="left">Edge Adjacency Indices Descriptors</td>
<td align="left">97</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAEigIdx</td>
<td align="left">Eigenvalue-Based Indices Descriptors</td>
<td align="left">44</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAFGC</td>
<td align="left">Functional Group Counts Descriptors</td>
<td align="left">5</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAGeom</td>
<td align="left">Geometrical Descriptors</td>
<td align="left">41</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAGETAWAY</td>
<td align="left">GETAWAY Descriptors</td>
<td align="left">194</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAInfo</td>
<td align="left">Information Indices Descriptors</td>
<td align="left">47</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAMolProp</td>
<td align="left">Molecular Properties Descriptors</td>
<td align="left">12</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AARandic</td>
<td align="left">Randic Molecular Profiles Descriptors</td>
<td align="left">41</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AARDF</td>
<td align="left">RDF Descriptors</td>
<td align="left">82</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AATopo</td>
<td align="left">Topological Descriptors</td>
<td align="left">78</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AATopoChg</td>
<td align="left">Topological Charge Indices Descriptors</td>
<td align="left">15</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAWalk</td>
<td align="left">Walk and Path Counts Descriptors</td>
<td align="left">40</td>
<td align="left">Dragon</td>
</tr>
<tr class="odd">
<td align="left">AAWHIM</td>
<td align="left">WHIM Descriptors</td>
<td align="left">99</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AACPSA</td>
<td align="left">CPSA Descriptors</td>
<td align="left">41</td>
<td align="left">Accelrys Discovery Studio</td>
</tr>
<tr class="odd">
<td align="left">AADescAll</td>
<td align="left">All the 2D Descriptors Calculated by Dragon</td>
<td align="left">1171</td>
<td align="left">Dragon</td>
</tr>
<tr class="even">
<td align="left">AAMOE2D</td>
<td align="left">All the 2D Descriptors Calculated by MOE</td>
<td align="left">148</td>
<td align="left">MOE</td>
</tr>
<tr class="odd">
<td align="left">AAMOE3D</td>
<td align="left">All the 3D Descriptors Calculated by MOE</td>
<td align="left">143</td>
<td align="left">MOE</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-choupaac" class="csl-entry">
Chou,K.-C. (2001) Prediction of protein cellular attributes using
pseudo-amino acid composition. <em>PROTEINS: Structure, Function, and
Genetics</em>, <strong>43</strong>, 246â€“255.
</div>
<div id="ref-chouqsoe" class="csl-entry">
Chou,K.-C. (2000) Prediction of protein subcellar locations by
incorporating quasi-sequence-order effect. <em>Biochemical and
Biophysical Research Communications</em>, <strong>278</strong>, 477â€“483.
</div>
<div id="ref-chouapaac" class="csl-entry">
Chou,K.-C. (2005) Using amphiphilic pseudo amino acid composition to
predict enzyme subfamily classes. <em>Bioinformatics</em>,
<strong>21</strong>, 10â€“19.
</div>
<div id="ref-chou2008cell" class="csl-entry">
Chou,K.-C. and Shen,H.-B. (2008) Cell-PLoc: A package of web servers for
predicting subcellular localization of proteins in various organisms.
<em>Nature Protocols</em>, <strong>3</strong>, 153â€“162.
</div>
<div id="ref-dubchak1" class="csl-entry">
Dubchak,I. <em>et al.</em> (1995) Prediction of protein folding class
using global description of amino acid sequence. <em>Proceedings of the
National Academy of Sciences</em>, <strong>92</strong>, 8700â€“8704.
</div>
<div id="ref-dubchak2" class="csl-entry">
Dubchak,I. <em>et al.</em> (1999) Recognition of a protein fold in the
context of the SCOP classification. <em>Proteins: Structure, Function
and Genetics</em>, <strong>35</strong>, 401â€“407.
</div>
<div id="ref-grantham" class="csl-entry">
Grantham,R. (1974) Amino acid difference formula to help explain protein
evolution. <em>Science</em>, <strong>185</strong>, 862â€“864.
</div>
<div id="ref-aaindex1999" class="csl-entry">
Kawashima,S. <em>et al.</em> (1999) AAindex: Amino acid index database.
<em>Nucleic Acids Research</em>, <strong>27</strong>, 368â€“369.
</div>
<div id="ref-aaindex2008" class="csl-entry">
Kawashima,S. <em>et al.</em> (2008) AAindex: Amino acid index database
(progress report). <em>Nucleic Acids Research</em>, <strong>36</strong>,
D202â€“D205.
</div>
<div id="ref-aaindex2000" class="csl-entry">
Kawashima,S. and Kanehisa,M. (2000) AAindex: Amino acid index database.
<em>Nucleic Acids Research</em>, <strong>28</strong>, 374.
</div>
<div id="ref-rangwala2005profile" class="csl-entry">
Rangwala,H. and Karypis,G. (2005) Profile-based direct kernels for
remote homology detection and fold recognition. <em>Bioinformatics</em>,
<strong>21</strong>, 4239â€“4247.
</div>
<div id="ref-wrede" class="csl-entry">
Schneider,G. and Wrede,P. (1994) The rational design of amino acid
sequences by artificial neural networks and simulated molecular
evolution: Do novo design of an idealized leader cleavage site.
<em>Biophysical Journal</em>, <strong>66</strong>, 335â€“344.
</div>
<div id="ref-shenjw" class="csl-entry">
Shen,J.W. <em>et al.</em> (2007) Predicting protein-protein interactions
based only on sequences information. <em>Proceedings of the National
Academy of Sciences</em>, <strong>104</strong>, 4337â€“4341.
</div>
<div id="ref-Xiao2015" class="csl-entry">
Xiao,N. <em>et al.</em> (2015) <a href="https://doi.org/10.1093/bioinformatics/btv042" class="external-link"><span class="nocase">protr/ProtrWeb: R package and web server for generating
various numerical representation schemes of protein
sequences</span></a>. <em>Bioinformatics</em>, <strong>31</strong>,
1857â€“1859.
</div>
<div id="ref-ye2011assessment" class="csl-entry">
Ye,X. <em>et al.</em> (2011) An assessment of substitution scores for
protein profileâ€“profile comparison. <em>Bioinformatics</em>,
<strong>27</strong>, 3356â€“3363.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Nan Xiao, Qing-Song Xu, Dong-Sheng Cao.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
